From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dan Carney <dcarney@cloudflare.com>
Date: Wed, 17 Dec 2025 10:59:18 +0000
Subject: disable info level logging, re2


diff --git a/include/perfetto/base/build_configs/bazel/perfetto_build_flags.h b/include/perfetto/base/build_configs/bazel/perfetto_build_flags.h
index a5519e4de613b99164450f1fd9f857e236b33537..9be7921eb078b1868d66d95c5c9f0e46460b45a2 100644
--- a/include/perfetto/base/build_configs/bazel/perfetto_build_flags.h
+++ b/include/perfetto/base/build_configs/bazel/perfetto_build_flags.h
@@ -37,5 +37,5 @@
 #define PERFETTO_BUILDFLAG_DEFINE_PERFETTO_FORCE_DCHECK_OFF() (1)
 #define PERFETTO_BUILDFLAG_DEFINE_PERFETTO_HTTP_ADDITIONAL_CORS_ORIGINS() ("")
-#define PERFETTO_BUILDFLAG_DEFINE_PERFETTO_VERBOSE_LOGS() (1)
+#define PERFETTO_BUILDFLAG_DEFINE_PERFETTO_VERBOSE_LOGS() (0)
 #define PERFETTO_BUILDFLAG_DEFINE_PERFETTO_VERSION_GEN() (1)
 #define PERFETTO_BUILDFLAG_DEFINE_PERFETTO_TP_PERCENTILE() (1)
@@ -57,5 +57,5 @@
 #define PERFETTO_BUILDFLAG_DEFINE_PERFETTO_ENABLE_LOCKFREE_TASKRUNNER() (1)
 #define PERFETTO_BUILDFLAG_DEFINE_PERFETTO_ENABLE_SOCK_INOTIFY() (1)
-#define PERFETTO_BUILDFLAG_DEFINE_PERFETTO_RE2() (1)
+#define PERFETTO_BUILDFLAG_DEFINE_PERFETTO_RE2() (0)
 #define PERFETTO_BUILDFLAG_DEFINE_PERFETTO_PCRE2() (0)
 #define PERFETTO_BUILDFLAG_DEFINE_PERFETTO_AMALGAMATED_SDK() (0)
