diff --git a/dist/shared/capnp-es.CBN2NTjq.mjs b/dist/shared/capnp-es.CBN2NTjq.mjs
index 663c15f2f2ec34621d8d1d7f30cf91245cb1a44a..a589b30bf0ae2af42081567a6b49357bb3881475 100644
--- a/dist/shared/capnp-es.CBN2NTjq.mjs
+++ b/dist/shared/capnp-es.CBN2NTjq.mjs
@@ -17,9 +17,9 @@ async function cliMain(outFormat) {
       dataBuf = await execCapnpc(sources, options, outDir);
     }
     const { files } = await compileAll(dataBuf, {
-      ts: outFormats.includes("ts"),
-      js: outFormats.includes("js"),
-      dts: outFormats.includes("dts")
+      ts: true,
+      js: true,
+      dts: false
     });
     (await writeFiles(files, outDir)).map(
       (file) => console.log(`[capnp-es] ${file}`)
