diff --git a/dist/compiler/main.mjs b/dist/compiler/main.mjs
index 6999c0e0c50570efd51959661ffc611f561bd31b..c37f97d5d786e4fc316acd4f5780e8e3976b21dd 100755
--- a/dist/compiler/main.mjs
+++ b/dist/compiler/main.mjs
@@ -38,7 +38,7 @@ Options:
 `;
 const sources = [];
 const options = [];
-let outFormats = ["js"];
+let outFormats = ["ts","js"];
 let outDir;
 for (const arg of process.argv.slice(2)) {
   if (arg === "--help") {
diff --git a/dist/shared/capnp-es.C3d1NAZK.mjs b/dist/shared/capnp-es.C3d1NAZK.mjs
index 08c44dc210d976e80cac085de9443068b367b2ff..61d6af8c4917d09ed7157a5098504eb28f85d729 100644
--- a/dist/shared/capnp-es.C3d1NAZK.mjs
+++ b/dist/shared/capnp-es.C3d1NAZK.mjs
@@ -2128,7 +2128,7 @@ async function compileAll(buff, opts) {
   }
   if (!opts?.ts) {
     for (const [fileName] of files) {
-      if (fileName.endsWith(".ts")) {
+      if (fileName.endsWith(".ts") && !fileName.endsWith(".d.ts")) {
         files.delete(fileName);
       }
     }
