diff --git a/MODULE.bazel b/MODULE.bazel
index a321128..cd672d2 100644
--- MODULE.bazel
+++ MODULE.bazel
@@ -12,6 +12,10 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.

+module(name = "perfetto")
+bazel_dep(name = "perfetto_cfg", version = "0.0.0")
+bazel_dep(name = "platforms", version = "1.0.0")
+bazel_dep(name = "protobuf", version = "33.1")
 bazel_dep(name = "rules_android", version = "0.6.0")

 android_sdk_repository_extension = use_extension(
@@ -43,4 +47,5 @@ android_ndk_repository_extension.configure(
 )
 use_repo(android_ndk_repository_extension, "androidndk")

-register_toolchains("@androidndk//:all")
+# Avoid eagerly referencing non-existent NDK when not doing Android development
+# register_toolchains("@androidndk//:all")
