diff -u bazel-spirv_tools_test/external/spirv_tools/build_defs.bzl /tmp/spirv_tools/build_defs.bzl
--- build_defs.bzl	2023-09-07 11:10:26.622407128 +0000
+++ build_defs.bzl	2023-09-07 11:00:44.410407128 +0000
@@ -8,7 +8,6 @@
     "//conditions:default": [
         "-DSPIRV_LINUX",
         "-DSPIRV_TIMER_ENABLED",
-        "-fvisibility=hidden",
         "-fno-exceptions",
         "-fno-rtti",
         "-Wall",
