From 37bc132e418017ff70e5c5a4562c457c2064f49b Mon Sep 17 00:00:00 2001
From: James M Snell <jasnell@gmail.com>
Date: Mon, 29 Apr 2024 13:22:55 -0700
Subject: Add iomanip


diff --git a/src/dawn/native/stream/ByteVectorSink.cpp b/src/dawn/native/stream/ByteVectorSink.cpp
--- a/src/dawn/native/stream/ByteVectorSink.cpp
+++ b/src/dawn/native/stream/ByteVectorSink.cpp
@@ -27,6 +27,7 @@

 #include "dawn/native/stream/ByteVectorSink.h"

+#include <iomanip>
 #include "dawn/native/stream/Stream.h"

 namespace dawn::native::stream {
