diff --git a/src/kj/compat/BUILD.bazel b/src/kj/compat/BUILD.bazel
index c21df55d..980429c6 100644
--- a/src/kj/compat/BUILD.bazel
+++ b/src/kj/compat/BUILD.bazel
@@ -48,6 +48,6 @@ cc_library(
     deps = [
         "//src/kj:kj-async-core",
         "//src/kj:kj-async-io",
-        "@zlib",
+        "@workerd//src/workerd/util:zlib",
     ],
 )
@@ -61,6 +61,6 @@ cc_library(
     deps = [
         "//src/kj:kj-async-core",
         "//src/kj:kj-async-io",
-        "@zlib",
+        "@workerd//src/workerd/util:zlib",
     ],
 )
diff --git a/src/kj/compat/gzip.h b/src/kj/compat/gzip.h
index 091251af..364370c0 100644
--- a/src/kj/compat/gzip.h
+++ b/src/kj/compat/gzip.h
@@ -24,3 +24,3 @@
+#include <workerd/util/zlib.h>
 #include <kj/io.h>
 #include <kj/async-io.h>
-#include <zlib.h>
diff --git a/src/kj/compat/http.c++ b/src/kj/compat/http.c++
index f3b34955..ac02c449 100644
--- a/src/kj/compat/http.c++
+++ b/src/kj/compat/http.c++
@@ -35 +35 @@
-#include <zlib.h>
+#include <workerd/util/zlib.h>
