diff -u5 -r sqlite-src-pristine-patched1/tool/mksqlite3c.tcl sqlite-src-modified/tool/mksqlite3c.tcl
--- sqlite-src-pristine-patched1/tool/mksqlite3c.tcl	2025-11-19 08:32:31.802107018 -0600
+++ sqlite-src-modified/tool/mksqlite3c.tcl	2025-11-19 08:28:09.082039967 -0600
@@ -87,11 +87,14 @@
 set fname sqlite3.c
 if {$enable_recover} { set fname sqlite3r.c }
 set out [open $fname wb]
 # Force the output to use unix line endings, even on Windows.
 fconfigure $out -translation binary
-set today [clock format [clock seconds] -format "%Y-%m-%d %H:%M:%S UTC" -gmt 1]
+# The command below results in "couldn't find HOME environment variable to
+# expand path" errors on macOS CI runs. today is unused, so it is safe to
+# comment it out.
+#set today [clock format [clock seconds] -format "%Y-%m-%d %H:%M:%S UTC" -gmt 1]
 puts $out [subst \
 {/******************************************************************************
 ** This file is an amalgamation of many separate C source files from SQLite
 ** version $VERSION.  By combining all the individual C code files into this
 ** single large file, the entire code can be compiled as a single translation
