diff -u5 -r sqlite-src-pristine/src/wal.c sqlite-src-modified/src/wal.c
--- sqlite-src-pristine/src/wal.c	2024-10-21 09:47:53.000000000 -0700
+++ sqlite-src-modified/src/wal.c	2025-06-17 23:10:43.657220118 -0700
@@ -3760,10 +3760,11 @@
         rc = xUndo(pUndoCtx, walFramePgno(pWal, iFrame));
       }
       if( iMax!=pWal->hdr.mxFrame ) walCleanupHash(pWal);
     }
     SEH_EXCEPT( rc = SQLITE_IOERR_IN_PAGE; )
+    pWal->iReCksum = 0;
   }
   return rc;
 }
 
 /*
@@ -3807,10 +3808,13 @@
     pWal->hdr.aFrameCksum[1] = aWalData[2];
     SEH_TRY {
       walCleanupHash(pWal);
     }
     SEH_EXCEPT( rc = SQLITE_IOERR_IN_PAGE; )
+    if( pWal->iReCksum>pWal->hdr.mxFrame ){
+      pWal->iReCksum = 0;
+    }
   }
 
   return rc;
 }
 
