From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dhi Aurrahman <dio@rockybars.com>
Date: Thu, 27 Oct 2022 12:45:05 +0700
Subject: Allow compiling on macOS catalina and ventura

Signed-off-by: Dhi Aurrahman <dio@rockybars.com>

diff --git a/bazel/defs.bzl b/bazel/defs.bzl
index 8c3e3b9d1d9a37679a36c61c1fc03342d7f25587..8053fa95df82eee6c3b2286e56e68129ec3c876f 100644
--- a/bazel/defs.bzl
+++ b/bazel/defs.bzl
@@ -151,6 +151,18 @@ def _default_args():
                 "-fno-integrated-as",
             ],
             "//conditions:default": [],
+        }) + select({
+            "@v8//bazel/config:is_macos": [
+                # The clang available on macOS Catalina has a warning that isn't clean on v8 code.
+                # The warning doesn't show up with more recent clangs, so disable it for now.
+                "-Wno-range-loop-analysis",
+
+                # To supress warning on deprecated declaration on v8 code:
+                # src/base/platform/platform-darwin.cc:56:22:
+                # 'getsectdatafromheader_64' is deprecated: first deprecated in macOS 13.0.
+                "-Wno-deprecated-declarations",
+            ],
+            "//conditions:default": [],
         }),
         cxxopts = select({
             "//third_party/v8/HEAD/google3/config:is_opt": [
