From 094a86545cd7a11e924cfd131bea6e7e6cb23a79 Mon Sep 17 00:00:00 2001
From: Dhi Aurrahman <dio@rockybars.com>
Date: Thu, 27 Oct 2022 12:45:05 +0700
Subject: Allow compiling on macOS catalina and ventura

Signed-off-by: Dhi Aurrahman <dio@rockybars.com>
---
 bazel/defs.bzl | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/bazel/defs.bzl b/bazel/defs.bzl
index 9024b1e14f6f032d636e26e18f88db6d3365ee4c..459979f1637199e06a9ba5b0df6d63c3d778a460 100644
--- a/bazel/defs.bzl
+++ b/bazel/defs.bzl
@@ -156,6 +156,18 @@ def _default_args():
                 "-fno-integrated-as",
             ],
             "//conditions:default": [],
+        }) + select({
+            "@v8//bazel/config:is_macos": [
+                # The clang available on macOS Catalina has a warning that isn't clean on v8 code.
+                # The warning doesn't show up with more recent clangs, so disable it for now.
+                "-Wno-range-loop-analysis",
+
+                # To supress warning on deprecated declaration on v8 code:
+                # src/base/platform/platform-darwin.cc:56:22:
+                # 'getsectdatafromheader_64' is deprecated: first deprecated in macOS 13.0.
+                "-Wno-deprecated-declarations",
+            ],
+            "//conditions:default": [],
         }) +  select({
             "@v8//bazel/config:is_opt_android": [
                 "-fvisibility=hidden",
