From 8831ef5a2250f102861c0c56bf9f626a7555a8e8 Mon Sep 17 00:00:00 2001
From: Felix Hanau <felix@cloudflare.com>
Date: Wed, 8 Mar 2023 16:15:31 -0500
Subject: Fix v8 code_generator imports

---
 BUILD.bazel | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/BUILD.bazel b/BUILD.bazel
index 83cb5a4d6b..366312ce07 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -3592,6 +3592,11 @@ py_binary(
     deps = [
         requirement("jinja2"),
     ],
+    # The code_generator complains about not finding pdl.py, so amend the list
+    # of include directories. Not sure why this is needed only with bazel v6.
+    imports = [
+        "third_party/inspector_protocol/",
+    ],
 )
 
 genrule(
