From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: James M Snell <jasnell@gmail.com>
Date: Thu, 22 Jun 2023 15:29:26 -0700
Subject: Implement Promise Context Tagging


diff --git a/include/v8-callbacks.h b/include/v8-callbacks.h
index e5eba5a203b8bc4d0c05b1f0d6cbdffd352d4a06..cfba4bb26f865c0e38574f796200ffc5e0dc60fc 100644
--- a/include/v8-callbacks.h
+++ b/include/v8-callbacks.h
@@ -528,6 +528,14 @@ using FilterETWSessionByURL2Callback = FilterETWSessionByURLResult (*)(
     Local<Context> context, const std::string& etw_filter_payload);
 #endif  // V8_OS_WIN
 
+/**
+ * PromiseCrossContextCallback is called when following a promise and the
+ * promise's context tag is not strictly equal to the isolate's current
+ * promise context tag.
+ */
+using PromiseCrossContextCallback = MaybeLocal<Promise> (*)(
+    Local<Context> context, Local<Promise> promise, Local<Object> tag);
+
 }  // namespace v8
 
 #endif  // INCLUDE_V8_ISOLATE_CALLBACKS_H_
diff --git a/include/v8-isolate.h b/include/v8-isolate.h
index 44bde532a6253f7c1891dbb51dc3de21daf7a238..8f620d08c0b8919fc3312c53bd9efa5d11ded1c6 100644
--- a/include/v8-isolate.h
+++ b/include/v8-isolate.h
@@ -1875,6 +1875,9 @@ class V8_EXPORT Isolate {
    */
   uint64_t GetHashSeed();
 
+  class PromiseContextScope;
+  void SetPromiseCrossContextCallback(PromiseCrossContextCallback callback);
+
   Isolate() = delete;
   ~Isolate() = delete;
   Isolate(const Isolate&) = delete;
@@ -1921,6 +1924,19 @@ MaybeLocal<T> Isolate::GetDataFromSnapshotOnce(size_t index) {
   return {};
 }
 
+class Isolate::PromiseContextScope {
+ public:
+  PromiseContextScope(Isolate* isolate, v8::Local<v8::Object> tag);
+  ~PromiseContextScope();
+  PromiseContextScope(const PromiseContextScope&) = delete;
+  PromiseContextScope(PromiseContextScope&&) = delete;
+  PromiseContextScope& operator=(const PromiseContextScope&) = delete;
+  PromiseContextScope& operator=(PromiseContextScope&&) = delete;
+
+ private:
+  internal::Isolate* isolate_;
+};
+
 }  // namespace v8
 
 #endif  // INCLUDE_V8_ISOLATE_H_
diff --git a/src/api/api.cc b/src/api/api.cc
index b169775e02902f1839f387b60b112d4b73e2725a..56dbb41b430b77cc69a9225a39a9de74c620bf0f 100644
--- a/src/api/api.cc
+++ b/src/api/api.cc
@@ -12692,6 +12692,25 @@ std::string SourceLocation::ToString() const {
       .str();
 }
 
+void Isolate::SetPromiseCrossContextCallback(
+    PromiseCrossContextCallback callback) {
+  i::Isolate* isolate = reinterpret_cast<i::Isolate*>(this);
+  isolate->set_promise_cross_context_callback(callback);
+}
+
+Isolate::PromiseContextScope::PromiseContextScope(Isolate* isolate,
+                                                  v8::Local<v8::Object> tag)
+    : isolate_(reinterpret_cast<i::Isolate*>(isolate)) {
+  DCHECK(!isolate_->has_promise_context_tag());
+  DCHECK(!tag.IsEmpty());
+  i::Handle<i::Object> handle = Utils::OpenHandle(*tag);
+  isolate_->set_promise_context_tag(*handle);
+}
+
+Isolate::PromiseContextScope::~PromiseContextScope() {
+  isolate_->clear_promise_context_tag();
+}
+
 }  // namespace v8
 
 #ifdef ENABLE_SLOW_DCHECKS
diff --git a/src/builtins/promise-abstract-operations.tq b/src/builtins/promise-abstract-operations.tq
index b95d45fa9ae170f341c5e160b91cc8dcba7fe2cd..59b8d8d5e243cf46a8093c76613ae2ce420e22e8 100644
--- a/src/builtins/promise-abstract-operations.tq
+++ b/src/builtins/promise-abstract-operations.tq
@@ -20,6 +20,9 @@ extern transitioning runtime PromiseResolveAfterResolved(
 
 extern transitioning runtime PromiseRejectEventFromStack(
     implicit context: Context)(JSPromise, JSAny): JSAny;
+
+extern transitioning runtime PromiseContextCheck(
+    implicit context: Context)(JSPromise): JSPromise;
 }
 
 // https://tc39.es/ecma262/#sec-promise-abstract-operations
@@ -470,12 +473,14 @@ transitioning macro PerformPromiseThenImpl(
     // PromiseReaction holding both the onFulfilled and onRejected callbacks.
     // Once the {promise} is resolved we decide on the concrete handler to
     // push onto the microtask queue.
+    const delegate = runtime::PromiseContextCheck(promise);
     const promiseReactions =
-        UnsafeCast<(Zero | PromiseReaction)>(promise.reactions_or_result);
+        UnsafeCast<(Zero | PromiseReaction)>(delegate.reactions_or_result);
 
     const reaction = NewPromiseReaction(
         promiseReactions, resultPromiseOrCapability, onFulfilled, onRejected);
-    promise.reactions_or_result = reaction;
+    delegate.reactions_or_result = reaction;
+    delegate.SetHasHandler();
   } else {
     const reactionsOrResult = promise.reactions_or_result;
     let microtask: PromiseReactionJobTask;
@@ -497,8 +502,8 @@ transitioning macro PerformPromiseThenImpl(
         }
       }
     EnqueueMicrotask(handlerContext, microtask);
+    promise.SetHasHandler();
   }
-  promise.SetHasHandler();
 }
 
 transitioning javascript builtin PerformPromiseThenFunction(
diff --git a/src/builtins/promise-constructor.tq b/src/builtins/promise-constructor.tq
index 50677631b5399453eebc6b149272431f74b1fce6..c652bd836b27805865e0a902ef9cf7c1ff254646 100644
--- a/src/builtins/promise-constructor.tq
+++ b/src/builtins/promise-constructor.tq
@@ -8,6 +8,9 @@
 namespace runtime {
 extern transitioning runtime PromiseHookInit(
     implicit context: Context)(Object, Object): JSAny;
+
+extern transitioning runtime PromiseContextInit(
+    implicit context: Context)(JSPromise): JSAny;
 }
 
 // https://tc39.es/ecma262/#sec-promise-constructor
@@ -62,6 +65,7 @@ transitioning javascript builtin PromiseConstructor(
     result = UnsafeCast<JSPromise>(
         FastNewObject(context, promiseFun, UnsafeCast<JSReceiver>(newTarget)));
     PromiseInit(result);
+    runtime::PromiseContextInit(result);
     RunAnyPromiseHookInit(result, Undefined);
   }
 
diff --git a/src/builtins/promise-misc.tq b/src/builtins/promise-misc.tq
index f83ee777f596738f1a71606ba61a3a7fdbc2cd30..abfa23d9d5087148a25261e8a4aefdfc37a4b228 100644
--- a/src/builtins/promise-misc.tq
+++ b/src/builtins/promise-misc.tq
@@ -53,6 +53,7 @@ macro PromiseInit(promise: JSPromise): void {
     is_silent: false,
     async_task_id: kInvalidAsyncTaskId
   });
+  promise.context_tag = kZero;
   promise_internal::ZeroOutEmbedderOffsets(promise);
 }
 
@@ -72,6 +73,7 @@ macro InnerNewJSPromise(implicit context: Context)(): JSPromise {
     is_silent: false,
     async_task_id: kInvalidAsyncTaskId
   });
+  promise.context_tag = kZero;
   return promise;
 }
 
@@ -271,6 +273,7 @@ transitioning macro NewJSPromise(implicit context: Context)(
                                  parent: Object): JSPromise {
   const instance = InnerNewJSPromise();
   PromiseInit(instance);
+  runtime::PromiseContextInit(instance);
   RunAnyPromiseHookInit(instance, parent);
   return instance;
 }
@@ -294,6 +297,7 @@ transitioning macro NewJSPromise(
   instance.reactions_or_result = result;
   instance.SetStatus(status);
   promise_internal::ZeroOutEmbedderOffsets(instance);
+  runtime::PromiseContextInit(instance);
   RunAnyPromiseHookInit(instance, Undefined);
   return instance;
 }
diff --git a/src/compiler/js-create-lowering.cc b/src/compiler/js-create-lowering.cc
index bbf3b14898eaa5caf9e90ff054048a3b197e15d5..56aa2697ba7377430222d09e149ce3bdc72ed116 100644
--- a/src/compiler/js-create-lowering.cc
+++ b/src/compiler/js-create-lowering.cc
@@ -1123,10 +1123,12 @@ Reduction JSCreateLowering::ReduceJSCreatePromise(Node* node) {
   a.Store(AccessBuilder::ForJSObjectOffset(
               offsetof(JSPromise, reactions_or_result_)),
           jsgraph()->ZeroConstant());
+  a.Store(AccessBuilder::ForJSObjectOffset(JSPromise::kContextTagOffset),
+          jsgraph()->ZeroConstant());
   static_assert(v8::Promise::kPending == 0);
   a.Store(AccessBuilder::ForJSObjectOffset(offsetof(JSPromise, flags_)),
           jsgraph()->ZeroConstant());
-  static_assert(sizeof(JSPromise) == 5 * kTaggedSize);
+  static_assert(sizeof(JSPromise) == 6 * kTaggedSize);
   for (int offset = static_cast<int>(sizeof(JSPromise));
        offset < static_cast<int>(sizeof(JSPromise)) +
                     v8::Promise::kEmbedderFieldCount * kEmbedderDataSlotSize;
diff --git a/src/diagnostics/objects-printer.cc b/src/diagnostics/objects-printer.cc
index 97d027774870112a7b9050c2cd8bc0f9dc27a971..974359132e4d6c4d7a0aa02fbfe7a1663bb483c7 100644
--- a/src/diagnostics/objects-printer.cc
+++ b/src/diagnostics/objects-printer.cc
@@ -1021,6 +1021,7 @@ void JSPromise::JSPromisePrint(std::ostream& os) {
   }
   os << "\n - has_handler: " << has_handler();
   os << "\n - is_silent: " << is_silent();
+  os << "\n - context_tag: " << Brief(context_tag());
   JSObjectPrintBody(os, this);
 }
 
diff --git a/src/execution/isolate-inl.h b/src/execution/isolate-inl.h
index 50b50e7517fa9683b484fc16bbcba309bcdaab3d..7b5988dc0cf5ceadac74136e61a3b20bcf0ac7c0 100644
--- a/src/execution/isolate-inl.h
+++ b/src/execution/isolate-inl.h
@@ -126,6 +126,25 @@ bool Isolate::is_execution_terminating() {
          i::ReadOnlyRoots(this).termination_exception();
 }
 
+Tagged<Object> Isolate::promise_context_tag() { return promise_context_tag_; }
+
+bool Isolate::has_promise_context_tag() {
+  return promise_context_tag_ != ReadOnlyRoots(this).the_hole_value();
+}
+
+void Isolate::clear_promise_context_tag() {
+  set_promise_context_tag(ReadOnlyRoots(this).the_hole_value());
+}
+
+void Isolate::set_promise_context_tag(Tagged<Object> tag) {
+  promise_context_tag_ = tag;
+}
+
+void Isolate::set_promise_cross_context_callback(
+    PromiseCrossContextCallback callback) {
+  promise_cross_context_callback_ = callback;
+}
+
 #ifdef DEBUG
 Tagged<Object> Isolate::VerifyBuiltinsResult(Tagged<Object> result) {
   if (is_execution_terminating() && !v8_flags.strict_termination_checks) {
diff --git a/src/execution/isolate.cc b/src/execution/isolate.cc
index 682c93049ee8c1776bbd1db323eaa4812d15ac83..fd8817a012c2221f35c442bbf4b092a86cb5c23b 100644
--- a/src/execution/isolate.cc
+++ b/src/execution/isolate.cc
@@ -681,6 +681,8 @@ void Isolate::Iterate(RootVisitor* v, ThreadLocalTop* thread) {
                       FullObjectSlot(&thread->pending_message_));
   v->VisitRootPointer(Root::kStackRoots, nullptr,
                       FullObjectSlot(&thread->context_));
+  v->VisitRootPointer(Root::kStackRoots, nullptr,
+                      FullObjectSlot(&promise_context_tag_));
 
   for (v8::TryCatch* block = thread->try_catch_handler_; block != nullptr;
        block = block->next_) {
@@ -6339,6 +6341,7 @@ bool Isolate::Init(SnapshotData* startup_snapshot_data,
     shared_heap_object_cache_.push_back(ReadOnlyRoots(this).undefined_value());
   }
 
+  clear_promise_context_tag();
   InitializeThreadLocal();
 
   // Profiler has to be created after ThreadLocal is initialized
@@ -8494,5 +8497,40 @@ void Isolate::PrintNumberStringCacheStats(const char* comment,
   PrintF("\n");
 }
 
+class Isolate::PromiseCrossContextCallbackScope {
+ public:
+  PromiseCrossContextCallbackScope(Isolate& isolate) : isolate_(isolate) {
+    DCHECK(!isolate_.in_promise_cross_context_callback_);
+    isolate_.in_promise_cross_context_callback_ = true;
+  }
+  ~PromiseCrossContextCallbackScope() {
+    isolate_.in_promise_cross_context_callback_ = false;
+  }
+
+ private:
+  Isolate& isolate_;
+};
+
+MaybeHandle<JSPromise> Isolate::RunPromiseCrossContextCallback(
+    Handle<NativeContext> context, Handle<JSPromise> promise) {
+  if (promise_cross_context_callback_ == nullptr ||
+      in_promise_cross_context_callback_) {
+    return promise;
+  }
+  PromiseCrossContextCallbackScope callback_scope(*this);
+  CHECK(IsJSReceiver(promise->context_tag()));
+
+  Handle<JSObject> context_tag(Cast<JSObject>(promise->context_tag()), this);
+  v8::Local<v8::Promise> result;
+  API_ASSIGN_RETURN_ON_EXCEPTION_VALUE(
+      this, result,
+      promise_cross_context_callback_(Utils::ToLocal(context),
+                                      v8::Utils::ToLocal(promise),
+                                      v8::Utils::ToLocal(context_tag)),
+      MaybeHandle<JSPromise>());
+
+  return v8::Utils::OpenHandle(*result);
+}
+
 }  // namespace internal
 }  // namespace v8
diff --git a/src/execution/isolate.h b/src/execution/isolate.h
index d0131fa4e09c8ba6e8ff7e92ae2a68dea9edcf4c..786652f5fe1d337aa92f89ea19f4c8feefea4ce2 100644
--- a/src/execution/isolate.h
+++ b/src/execution/isolate.h
@@ -2466,6 +2466,15 @@ class V8_EXPORT_PRIVATE Isolate final : private HiddenFactory {
                                        v8::ExceptionContext callback_kind);
   void SetExceptionPropagationCallback(ExceptionPropagationCallback callback);
 
+  inline Tagged<Object> promise_context_tag();
+  inline bool has_promise_context_tag();
+  inline void clear_promise_context_tag();
+  inline void set_promise_context_tag(Tagged<Object> tag);
+  inline void set_promise_cross_context_callback(
+      PromiseCrossContextCallback callback);
+  MaybeHandle<JSPromise> RunPromiseCrossContextCallback(
+      Handle<NativeContext> context, Handle<JSPromise> promise);
+
 #ifdef V8_ENABLE_WASM_SIMD256_REVEC
   void set_wasm_revec_verifier_for_test(
       compiler::turboshaft::WasmRevecVerifier* verifier) {
@@ -3001,6 +3010,12 @@ class V8_EXPORT_PRIVATE Isolate final : private HiddenFactory {
 
   bool is_frozen_ = false;
 
+  Tagged<Object> promise_context_tag_;
+  PromiseCrossContextCallback promise_cross_context_callback_;
+  bool in_promise_cross_context_callback_ = false;
+
+  class PromiseCrossContextCallbackScope;
+
   friend class GlobalSafepoint;
   friend class heap::HeapTester;
   friend class IsolateForPointerCompression;
@@ -3008,6 +3023,7 @@ class V8_EXPORT_PRIVATE Isolate final : private HiddenFactory {
   friend class IsolateGroup;
   friend class TestSerializer;
   friend class SharedHeapNoClientsTest;
+  friend class PromiseCrossContextCallbackScope;
 };
 
 // The current entered Isolate and its thread data. Do not access these
diff --git a/src/heap/factory.cc b/src/heap/factory.cc
index 5afe0042de2cb045ff86ce4ed380c2dc841a568d..ae04a98df1ed8a290095324b5daeff9146b73686 100644
--- a/src/heap/factory.cc
+++ b/src/heap/factory.cc
@@ -5040,6 +5040,12 @@ Handle<JSPromise> Factory::NewJSPromiseWithoutHook() {
   DisallowGarbageCollection no_gc;
   Tagged<JSPromise> raw = *promise;
   raw->set_reactions_or_result(Smi::zero(), SKIP_WRITE_BARRIER);
+  if (!isolate()->has_promise_context_tag()) {
+    raw->set_context_tag(Smi::zero(), SKIP_WRITE_BARRIER);
+  } else {
+    raw->set_context_tag(isolate()->promise_context_tag());
+  }
+
   raw->set_flags(0);
   // TODO(v8) remove once embedder data slots are always zero-initialized.
   InitEmbedderFields(*promise, Smi::zero());
diff --git a/src/maglev/maglev-graph-builder.cc b/src/maglev/maglev-graph-builder.cc
index 640cbd7ef0ae9a65462b558cd7297bef528231f2..4d353e7eda734e72ab67197cc858704d2b847a86 100644
--- a/src/maglev/maglev-graph-builder.cc
+++ b/src/maglev/maglev-graph-builder.cc
@@ -15379,9 +15379,10 @@ VirtualObject* MaglevGraphBuilder::CreateJSPromiseObject() {
   vobj->set(JSPromise::kElementsOffset,
             GetRootConstant(RootIndex::kEmptyFixedArray));
   vobj->set(offsetof(JSPromise, reactions_or_result_), GetSmiConstant(0));
+  vobj->set(JSPromise::kContextTagOffset, GetSmiConstant(0));
   static_assert(v8::Promise::kPending == 0);
   vobj->set(offsetof(JSPromise, flags_), GetSmiConstant(0));
-  static_assert(sizeof(JSPromise) == 5 * kTaggedSize);
+  static_assert(sizeof(JSPromise) == 6 * kTaggedSize);
   for (int offset = sizeof(JSPromise);
        offset < static_cast<int>(sizeof(JSPromise)) +
                     v8::Promise::kEmbedderFieldCount * kEmbedderDataSlotSize;
diff --git a/src/objects/js-promise-inl.h b/src/objects/js-promise-inl.h
index 21dfbffe3795544efb54d1b01dff2925c6de82f2..fc47da509721868f88fbfa0da566b8367d9db354 100644
--- a/src/objects/js-promise-inl.h
+++ b/src/objects/js-promise-inl.h
@@ -27,6 +27,12 @@ void JSPromise::set_reactions_or_result(
   reactions_or_result_.store(this, value, mode);
 }
 
+Tagged<Object> JSPromise::context_tag() const { return context_tag_.load(); }
+
+void JSPromise::set_context_tag(Tagged<Object> value, WriteBarrierMode mode) {
+  context_tag_.store(this, value, mode);
+}
+
 int JSPromise::flags() const { return flags_.load().value(); }
 
 void JSPromise::set_flags(int value) {
diff --git a/src/objects/js-promise.h b/src/objects/js-promise.h
index 19e3f89938b04136972d82ee66d8ff8f2c2447b5..fad6740ef43573befce77eb54053f5a43b3bf2b6 100644
--- a/src/objects/js-promise.h
+++ b/src/objects/js-promise.h
@@ -38,6 +38,10 @@ V8_OBJECT class JSPromise : public JSObjectWithEmbedderSlots {
       Tagged<UnionOf<PromiseReaction, JSAny>> value,
       WriteBarrierMode mode = UPDATE_WRITE_BARRIER);
 
+  inline Tagged<Object> context_tag() const;
+  inline void set_context_tag(Tagged<Object> value,
+                              WriteBarrierMode mode = UPDATE_WRITE_BARRIER);
+
   inline int flags() const;
   inline void set_flags(int value);
 
@@ -105,9 +109,16 @@ V8_OBJECT class JSPromise : public JSObjectWithEmbedderSlots {
   // Smi 0 terminated list of PromiseReaction objects in case the JSPromise
   // was not settled yet, otherwise the result.
   TaggedMember<UnionOf<PromiseReaction, JSAny>> reactions_or_result_;
+  // The context tag (workerd Promise Context Tagging extension).
+  TaggedMember<Object> context_tag_;
   // SmiTagged<JSPromiseFlags>.
   TaggedMember<Smi> flags_;
 
+  // Back-compat offset constant for the workerd Promise Context Tagging
+  // extension. Defined after the class body, like JSRegExp::kFlagsOffset etc.
+  static const int kContextTagOffset;
+
+
  private:
   // https://tc39.es/ecma262/#sec-triggerpromisereactions
   static Handle<Object> TriggerPromiseReactions(Isolate* isolate,
@@ -116,6 +127,9 @@ V8_OBJECT class JSPromise : public JSObjectWithEmbedderSlots {
                                                 PromiseReaction::Type type);
 } V8_OBJECT_END;
 
+inline constexpr int JSPromise::kContextTagOffset =
+    offsetof(JSPromise, context_tag_);
+
 }  // namespace internal
 }  // namespace v8
 
diff --git a/src/objects/js-promise.tq b/src/objects/js-promise.tq
index 11c4aff5cd69699aa6813498478962005c302e9f..532e163d98a5bc42658824ecdc0fdab96cc48b68 100644
--- a/src/objects/js-promise.tq
+++ b/src/objects/js-promise.tq
@@ -33,6 +33,7 @@ extern class JSPromise extends JSObjectWithEmbedderSlots {
   // Smi 0 terminated list of PromiseReaction objects in case the JSPromise was
   // not settled yet, otherwise the result.
   reactions_or_result: Zero|PromiseReaction|JSAny;
+  context_tag: Object;
   flags: SmiTagged<JSPromiseFlags>;
 }
 
diff --git a/src/profiler/heap-snapshot-generator.cc b/src/profiler/heap-snapshot-generator.cc
index 8edad48dcf13067ae838e8ffe9372c9da4a754d1..91f9ea62d2d0a94fdf5fafdf688e0c8f3a1b2a3b 100644
--- a/src/profiler/heap-snapshot-generator.cc
+++ b/src/profiler/heap-snapshot-generator.cc
@@ -2238,6 +2238,8 @@ void V8HeapExplorer::ExtractJSPromiseReferences(HeapEntry* entry,
   SetInternalReference(entry, "reactions_or_result",
                        promise->reactions_or_result(),
                        offsetof(JSPromise, reactions_or_result_));
+  SetInternalReference(entry, "context_tag", promise->context_tag(),
+                       JSPromise::kContextTagOffset);
 }
 
 void V8HeapExplorer::ExtractJSGeneratorObjectReferences(
diff --git a/src/runtime/runtime-promise.cc b/src/runtime/runtime-promise.cc
index cbe68d70430188fceab54bf3911c5d617e76cd62..896bac667ce40ef23c8c4fcd6174fcd2ebc2076f 100644
--- a/src/runtime/runtime-promise.cc
+++ b/src/runtime/runtime-promise.cc
@@ -240,5 +240,41 @@ RUNTIME_FUNCTION(Runtime_ConstructSuppressedError) {
   return *result;
 }
 
+RUNTIME_FUNCTION(Runtime_PromiseContextInit) {
+  HandleScope scope(isolate);
+  DCHECK_EQ(1, args.length());
+  if (!isolate->has_promise_context_tag()) {
+    args.at<JSPromise>(0)->set_context_tag(Smi::zero());
+  } else {
+    CHECK(!IsUndefined(isolate->promise_context_tag()));
+    args.at<JSPromise>(0)->set_context_tag(isolate->promise_context_tag());
+  }
+  return ReadOnlyRoots(isolate).undefined_value();
+}
+
+RUNTIME_FUNCTION(Runtime_PromiseContextCheck) {
+  HandleScope scope(isolate);
+  DCHECK_EQ(1, args.length());
+
+  Handle<JSPromise> promise = args.at<JSPromise>(0);
+
+  // If promise.context_tag() is strict equal to isolate.promise_context_tag(),
+  // or if the promise being checked does not have a context tag, we'll just
+  // return promise directly.
+  Tagged<Object> obj = promise->context_tag();
+  if (obj == Smi::zero() || obj == isolate->promise_context_tag()) {
+    return *promise;
+  }
+
+  // Otherwise we defer to the PromiseCrossContextCallback. If the callback
+  // has not been set, then it should just return the same promise back here.
+  Handle<JSPromise> result;
+  ASSIGN_RETURN_FAILURE_ON_EXCEPTION(isolate, result,
+                                     isolate->RunPromiseCrossContextCallback(
+                                         isolate->native_context(), promise));
+
+  return *result;
+}
+
 }  // namespace internal
 }  // namespace v8
diff --git a/src/runtime/runtime.h b/src/runtime/runtime.h
index 098819e04b21e838b7ed8d03c1897f585bc78444..d91af102ab39d4b4355181bb5cf525a64d3f64d0 100644
--- a/src/runtime/runtime.h
+++ b/src/runtime/runtime.h
@@ -434,20 +434,22 @@ constexpr bool CanTriggerGC(T... properties) {
   F(StrictNotEqual, 2, 1)                  \
   F(ReferenceEqual, 2, 1)
 
-#define FOR_EACH_INTRINSIC_PROMISE(F, I) \
-  F(EnqueueMicrotask, 1, 1)              \
-  F(PromiseHookAfter, 1, 1)              \
-  F(PromiseHookBefore, 1, 1)             \
-  F(PromiseHookInit, 2, 1)               \
-  F(PromiseRejectEventFromStack, 2, 1)   \
-  F(PromiseRevokeReject, 1, 1)           \
-  F(RejectPromise, 3, 1)                 \
-  F(ResolvePromise, 2, 1)                \
-  F(PromiseRejectAfterResolved, 2, 1)    \
-  F(PromiseResolveAfterResolved, 2, 1)   \
-  F(ConstructSuppressedError, 3, 1)      \
-  F(ConstructAggregateErrorHelper, 4, 1) \
-  F(ConstructInternalAggregateErrorHelper, -1 /* <= 5*/, 1)
+#define FOR_EACH_INTRINSIC_PROMISE(F, I)                    \
+  F(EnqueueMicrotask, 1, 1)                                 \
+  F(PromiseHookAfter, 1, 1)                                 \
+  F(PromiseHookBefore, 1, 1)                                \
+  F(PromiseHookInit, 2, 1)                                  \
+  F(PromiseRejectEventFromStack, 2, 1)                      \
+  F(PromiseRevokeReject, 1, 1)                              \
+  F(RejectPromise, 3, 1)                                    \
+  F(ResolvePromise, 2, 1)                                   \
+  F(PromiseRejectAfterResolved, 2, 1)                       \
+  F(PromiseResolveAfterResolved, 2, 1)                      \
+  F(ConstructSuppressedError, 3, 1)                         \
+  F(ConstructAggregateErrorHelper, 4, 1)                    \
+  F(ConstructInternalAggregateErrorHelper, -1 /* <= 5*/, 1) \
+  F(PromiseContextInit, 1, 1)                               \
+  F(PromiseContextCheck, 1, 1)
 
 #define FOR_EACH_INTRINSIC_PROXY(F, I) \
   F(CheckProxyGetSetTrapResult, 2, 1)  \
