From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: James M Snell <jasnell@gmail.com>
Date: Thu, 22 Jun 2023 15:29:26 -0700
Subject: Implement Promise Context Tagging


diff --git a/include/v8-callbacks.h b/include/v8-callbacks.h
index 456af07fe1d89feeb03daabc869072220808460d..2feae234f77b60de3220804edc6d2ccd067f5670 100644
--- a/include/v8-callbacks.h
+++ b/include/v8-callbacks.h
@@ -532,6 +532,14 @@ using FilterETWSessionByURL2Callback = FilterETWSessionByURLResult (*)(
     Local<Context> context, const std::string& etw_filter_payload);
 #endif  // V8_OS_WIN
 
+/**
+ * PromiseCrossContextCallback is called when following a promise and the
+ * promise's context tag is not strictly equal to the isolate's current
+ * promise context tag.
+ */
+using PromiseCrossContextCallback = MaybeLocal<Promise> (*)(
+    Local<Context> context, Local<Promise> promise, Local<Object> tag);
+
 }  // namespace v8
 
 #endif  // INCLUDE_V8_ISOLATE_CALLBACKS_H_
diff --git a/include/v8-isolate.h b/include/v8-isolate.h
index f1e39751b808e22fb7f134d6f606db8467d4d17d..523a60774c5cc3419256068f768213ccfb2ac8f0 100644
--- a/include/v8-isolate.h
+++ b/include/v8-isolate.h
@@ -1884,6 +1884,9 @@ class V8_EXPORT Isolate {
    */
   uint64_t GetHashSeed();
 
+  class PromiseContextScope;
+  void SetPromiseCrossContextCallback(PromiseCrossContextCallback callback);
+
   Isolate() = delete;
   ~Isolate() = delete;
   Isolate(const Isolate&) = delete;
@@ -1930,6 +1933,19 @@ MaybeLocal<T> Isolate::GetDataFromSnapshotOnce(size_t index) {
   return {};
 }
 
+class Isolate::PromiseContextScope {
+ public:
+  PromiseContextScope(Isolate* isolate, v8::Local<v8::Object> tag);
+  ~PromiseContextScope();
+  PromiseContextScope(const PromiseContextScope&) = delete;
+  PromiseContextScope(PromiseContextScope&&) = delete;
+  PromiseContextScope& operator=(const PromiseContextScope&) = delete;
+  PromiseContextScope& operator=(PromiseContextScope&&) = delete;
+
+ private:
+  internal::Isolate* isolate_;
+};
+
 }  // namespace v8
 
 #endif  // INCLUDE_V8_ISOLATE_H_
diff --git a/src/api/api.cc b/src/api/api.cc
index 669ccb3e90b6aba5ab30f8333b8ff2334e63e330..77ed046732334820fd0ce9c26a0c2905ac2ad702 100644
--- a/src/api/api.cc
+++ b/src/api/api.cc
@@ -12765,6 +12765,25 @@ std::string SourceLocation::ToString() const {
       .str();
 }
 
+void Isolate::SetPromiseCrossContextCallback(
+    PromiseCrossContextCallback callback) {
+  i::Isolate* isolate = reinterpret_cast<i::Isolate*>(this);
+  isolate->set_promise_cross_context_callback(callback);
+}
+
+Isolate::PromiseContextScope::PromiseContextScope(Isolate* isolate,
+                                                  v8::Local<v8::Object> tag)
+    : isolate_(reinterpret_cast<i::Isolate*>(isolate)) {
+  DCHECK(!isolate_->has_promise_context_tag());
+  DCHECK(!tag.IsEmpty());
+  i::Handle<i::Object> handle = Utils::OpenHandle(*tag);
+  isolate_->set_promise_context_tag(*handle);
+}
+
+Isolate::PromiseContextScope::~PromiseContextScope() {
+  isolate_->clear_promise_context_tag();
+}
+
 }  // namespace v8
 
 #ifdef ENABLE_SLOW_DCHECKS
diff --git a/src/builtins/promise-abstract-operations.tq b/src/builtins/promise-abstract-operations.tq
index e3a3ee118c865bd307a94dba39638676517ef6c0..b3b141983756253ec32f20bbe97c9bde1b1a7725 100644
--- a/src/builtins/promise-abstract-operations.tq
+++ b/src/builtins/promise-abstract-operations.tq
@@ -14,6 +14,9 @@ extern transitioning runtime PromiseRevokeReject(
 
 extern transitioning runtime PromiseRejectEventFromStack(
     implicit context: Context)(JSPromise, JSAny): JSAny;
+
+extern transitioning runtime PromiseContextCheck(
+    implicit context: Context)(JSPromise): JSPromise;
 }
 
 // https://tc39.es/ecma262/#sec-promise-abstract-operations
@@ -475,12 +478,14 @@ transitioning macro PerformPromiseThenImpl(
     // PromiseReaction holding both the onFulfilled and onRejected callbacks.
     // Once the {promise} is resolved we decide on the concrete handler to
     // push onto the microtask queue.
+    const delegate = runtime::PromiseContextCheck(promise);
     const promiseReactions =
-        UnsafeCast<(Zero | PromiseReaction)>(promise.reactions_or_result);
+        UnsafeCast<(Zero | PromiseReaction)>(delegate.reactions_or_result);
 
     const reaction = NewPromiseReaction(
         promiseReactions, resultPromiseOrCapability, onFulfilled, onRejected);
-    promise.reactions_or_result = reaction;
+    delegate.reactions_or_result = reaction;
+    delegate.SetHasHandler();
   } else {
     const reactionsOrResult = promise.reactions_or_result;
     let microtask: PromiseReactionJobTask;
@@ -502,8 +507,8 @@ transitioning macro PerformPromiseThenImpl(
         }
       }
     EnqueueMicrotask(handlerContext, microtask);
+    promise.SetHasHandler();
   }
-  promise.SetHasHandler();
 }
 
 transitioning javascript builtin PerformPromiseThenFunction(
diff --git a/src/builtins/promise-constructor.tq b/src/builtins/promise-constructor.tq
index 50677631b5399453eebc6b149272431f74b1fce6..c652bd836b27805865e0a902ef9cf7c1ff254646 100644
--- a/src/builtins/promise-constructor.tq
+++ b/src/builtins/promise-constructor.tq
@@ -8,6 +8,9 @@
 namespace runtime {
 extern transitioning runtime PromiseHookInit(
     implicit context: Context)(Object, Object): JSAny;
+
+extern transitioning runtime PromiseContextInit(
+    implicit context: Context)(JSPromise): JSAny;
 }
 
 // https://tc39.es/ecma262/#sec-promise-constructor
@@ -62,6 +65,7 @@ transitioning javascript builtin PromiseConstructor(
     result = UnsafeCast<JSPromise>(
         FastNewObject(context, promiseFun, UnsafeCast<JSReceiver>(newTarget)));
     PromiseInit(result);
+    runtime::PromiseContextInit(result);
     RunAnyPromiseHookInit(result, Undefined);
   }
 
diff --git a/src/builtins/promise-misc.tq b/src/builtins/promise-misc.tq
index eefb2d0c40d2fd98cbd1793e63841294aa28386d..4b9fb74a9952ed1216874319e2d29ea135c8207e 100644
--- a/src/builtins/promise-misc.tq
+++ b/src/builtins/promise-misc.tq
@@ -54,6 +54,7 @@ macro PromiseInit(promise: JSPromise): void {
     is_silent: false,
     async_task_id: kInvalidAsyncTaskId
   });
+  promise.context_tag = kZero;
   promise_internal::ZeroOutEmbedderOffsets(promise);
 }
 
@@ -74,6 +75,7 @@ macro InnerNewJSPromise(implicit context: Context)(): JSPromise {
     is_silent: false,
     async_task_id: kInvalidAsyncTaskId
   });
+  promise.context_tag = kZero;
   return promise;
 }
 
@@ -273,6 +275,7 @@ transitioning macro NewJSPromise(implicit context: Context)(parent: Object):
     JSPromise {
   const instance = InnerNewJSPromise();
   PromiseInit(instance);
+  runtime::PromiseContextInit(instance);
   RunAnyPromiseHookInit(instance, parent);
   return instance;
 }
@@ -296,6 +299,7 @@ transitioning macro NewJSPromise(
   instance.reactions_or_result = result;
   instance.SetStatus(status);
   promise_internal::ZeroOutEmbedderOffsets(instance);
+  runtime::PromiseContextInit(instance);
   RunAnyPromiseHookInit(instance, Undefined);
   return instance;
 }
diff --git a/src/compiler/js-create-lowering.cc b/src/compiler/js-create-lowering.cc
index 2a1946aea904042cbafe60d1afab46eaee755d1e..0851372fdc5c1c5876f35cea0d26782396249d84 100644
--- a/src/compiler/js-create-lowering.cc
+++ b/src/compiler/js-create-lowering.cc
@@ -1127,10 +1127,12 @@ Reduction JSCreateLowering::ReduceJSCreatePromise(Node* node) {
   a.Store(AccessBuilder::ForJSObjectOffset(
               offsetof(JSPromise, reactions_or_result_)),
           jsgraph()->ZeroConstant());
+  a.Store(AccessBuilder::ForJSObjectOffset(JSPromise::kContextTagOffset),
+          jsgraph()->ZeroConstant());
   static_assert(v8::Promise::kPending == 0);
   a.Store(AccessBuilder::ForJSObjectOffset(offsetof(JSPromise, flags_)),
           jsgraph()->ZeroConstant());
-  static_assert(sizeof(JSPromise) == 5 * kTaggedSize);
+  static_assert(sizeof(JSPromise) == 6 * kTaggedSize);
   for (int offset = static_cast<int>(sizeof(JSPromise));
        offset < static_cast<int>(sizeof(JSPromise)) +
                     v8::Promise::kEmbedderFieldCount * kEmbedderDataSlotSize;
diff --git a/src/diagnostics/objects-printer.cc b/src/diagnostics/objects-printer.cc
index 08aa74c0656176618057d8e53e4611c16c09e201..e52a24d496a419ac0d8f97135b6fcb098e57e778 100644
--- a/src/diagnostics/objects-printer.cc
+++ b/src/diagnostics/objects-printer.cc
@@ -1029,6 +1029,7 @@ void JSPromise::JSPromisePrint(std::ostream& os) {
   }
   os << "\n - has_handler: " << has_handler();
   os << "\n - is_silent: " << is_silent();
+  os << "\n - context_tag: " << Brief(context_tag());
   JSObjectPrintBody(os, this);
 }
 
diff --git a/src/execution/isolate-inl.h b/src/execution/isolate-inl.h
index de710c8fad334d5fc29ee0d4627b4eb370e58550..fb7d83dd6fcff7aa7aba2ba5ad533fa8b877e1e1 100644
--- a/src/execution/isolate-inl.h
+++ b/src/execution/isolate-inl.h
@@ -129,6 +129,25 @@ bool Isolate::is_javascript_execution_allowed() const {
          javascript_execution_dump();
 }
 
+Tagged<Object> Isolate::promise_context_tag() { return promise_context_tag_; }
+
+bool Isolate::has_promise_context_tag() {
+  return promise_context_tag_ != ReadOnlyRoots(this).the_hole_value();
+}
+
+void Isolate::clear_promise_context_tag() {
+  set_promise_context_tag(ReadOnlyRoots(this).the_hole_value());
+}
+
+void Isolate::set_promise_context_tag(Tagged<Object> tag) {
+  promise_context_tag_ = tag;
+}
+
+void Isolate::set_promise_cross_context_callback(
+    PromiseCrossContextCallback callback) {
+  promise_cross_context_callback_ = callback;
+}
+
 #ifdef DEBUG
 Tagged<Object> Isolate::VerifyBuiltinsResult(Tagged<Object> result) {
   if (is_execution_terminating() && !v8_flags.strict_termination_checks) {
diff --git a/src/execution/isolate.cc b/src/execution/isolate.cc
index ca7c4aab5a6df4d6c7eba37f345124e00f40f108..b4c92ae639b45b412185fecff2a78206f0422866 100644
--- a/src/execution/isolate.cc
+++ b/src/execution/isolate.cc
@@ -681,6 +681,8 @@ void Isolate::Iterate(RootVisitor* v, ThreadLocalTop* thread) {
                       FullObjectSlot(&thread->context_));
   v->VisitRootPointer(Root::kStackRoots, nullptr,
                       FullObjectSlot(&thread->last_entered_context_));
+  v->VisitRootPointer(Root::kStackRoots, nullptr,
+                      FullObjectSlot(&promise_context_tag_));
 
   for (v8::TryCatch* block = thread->try_catch_handler_; block != nullptr;
        block = block->next_) {
@@ -6498,6 +6500,7 @@ bool Isolate::Init(SnapshotData* startup_snapshot_data,
     shared_heap_object_cache_.push_back(ReadOnlyRoots(this).undefined_value());
   }
 
+  clear_promise_context_tag();
   InitializeThreadLocal();
 
   // Profiler has to be created after ThreadLocal is initialized
@@ -8712,5 +8715,40 @@ void Isolate::PrintNumberStringCacheStats(const char* comment,
   PrintF("\n");
 }
 
+class Isolate::PromiseCrossContextCallbackScope {
+ public:
+  PromiseCrossContextCallbackScope(Isolate& isolate) : isolate_(isolate) {
+    DCHECK(!isolate_.in_promise_cross_context_callback_);
+    isolate_.in_promise_cross_context_callback_ = true;
+  }
+  ~PromiseCrossContextCallbackScope() {
+    isolate_.in_promise_cross_context_callback_ = false;
+  }
+
+ private:
+  Isolate& isolate_;
+};
+
+MaybeHandle<JSPromise> Isolate::RunPromiseCrossContextCallback(
+    Handle<NativeContext> context, Handle<JSPromise> promise) {
+  if (promise_cross_context_callback_ == nullptr ||
+      in_promise_cross_context_callback_) {
+    return promise;
+  }
+  PromiseCrossContextCallbackScope callback_scope(*this);
+  CHECK(IsJSReceiver(promise->context_tag()));
+
+  Handle<JSObject> context_tag(Cast<JSObject>(promise->context_tag()), this);
+  v8::Local<v8::Promise> result;
+  API_ASSIGN_RETURN_ON_EXCEPTION_VALUE(
+      this, result,
+      promise_cross_context_callback_(Utils::ToLocal(context),
+                                      v8::Utils::ToLocal(promise),
+                                      v8::Utils::ToLocal(context_tag)),
+      MaybeHandle<JSPromise>());
+
+  return v8::Utils::OpenHandle(*result);
+}
+
 }  // namespace internal
 }  // namespace v8
diff --git a/src/execution/isolate.h b/src/execution/isolate.h
index 51c0f56f391a841f19a870a4a468da115fb9143e..22b3c246daa5b5bfc84129868dea110b426a476d 100644
--- a/src/execution/isolate.h
+++ b/src/execution/isolate.h
@@ -2512,6 +2512,15 @@ class V8_EXPORT_PRIVATE Isolate final : private HiddenFactory {
                                        v8::ExceptionContext callback_kind);
   void SetExceptionPropagationCallback(ExceptionPropagationCallback callback);
 
+  inline Tagged<Object> promise_context_tag();
+  inline bool has_promise_context_tag();
+  inline void clear_promise_context_tag();
+  inline void set_promise_context_tag(Tagged<Object> tag);
+  inline void set_promise_cross_context_callback(
+      PromiseCrossContextCallback callback);
+  MaybeHandle<JSPromise> RunPromiseCrossContextCallback(
+      Handle<NativeContext> context, Handle<JSPromise> promise);
+
 #ifdef V8_ENABLE_WASM_SIMD256_REVEC
   void set_wasm_revec_verifier_for_test(
       compiler::turboshaft::WasmRevecVerifier* verifier) {
@@ -3054,6 +3063,12 @@ class V8_EXPORT_PRIVATE Isolate final : private HiddenFactory {
   bool is_frozen_ = false;
 
   friend class HandleScopeImplementer;
+  Tagged<Object> promise_context_tag_;
+  PromiseCrossContextCallback promise_cross_context_callback_;
+  bool in_promise_cross_context_callback_ = false;
+
+  class PromiseCrossContextCallbackScope;
+
   friend class GlobalSafepoint;
   friend class heap::HeapTester;
   friend class IsolateForPointerCompression;
@@ -3061,6 +3076,7 @@ class V8_EXPORT_PRIVATE Isolate final : private HiddenFactory {
   friend class IsolateGroup;
   friend class TestSerializer;
   friend class SharedHeapNoClientsTest;
+  friend class PromiseCrossContextCallbackScope;
 };
 
 // The current entered Isolate and its thread data. Do not access these
diff --git a/src/heap/factory.cc b/src/heap/factory.cc
index 9716d02960b59ccb587e8549c4b223e6bb8b886f..931f8e0cb36010b83a3f27c384948fb1831e6118 100644
--- a/src/heap/factory.cc
+++ b/src/heap/factory.cc
@@ -5128,6 +5128,12 @@ Handle<JSPromise> Factory::NewJSPromiseWithoutHook() {
   DisallowGarbageCollection no_gc;
   Tagged<JSPromise> raw = *promise;
   raw->set_reactions_or_result(Smi::zero(), SKIP_WRITE_BARRIER);
+  if (!isolate()->has_promise_context_tag()) {
+    raw->set_context_tag(Smi::zero(), SKIP_WRITE_BARRIER);
+  } else {
+    raw->set_context_tag(isolate()->promise_context_tag());
+  }
+
   raw->set_flags(0);
   // TODO(v8) remove once embedder data slots are always zero-initialized.
   InitEmbedderFields(*promise, Smi::zero());
diff --git a/src/maglev/maglev-reducer-inl.h b/src/maglev/maglev-reducer-inl.h
index 61f857a3a0afc7ace8ca60b45b641dc5d6bc0b06..14492f4d2f99ea1156e454dd34862ae5f7c25cc6 100644
--- a/src/maglev/maglev-reducer-inl.h
+++ b/src/maglev/maglev-reducer-inl.h
@@ -5881,9 +5881,10 @@ VirtualObject* MaglevReducer<BaseT>::CreateJSPromiseObject() {
   vobj->set(offsetof(JSObject, elements_),
             GetRootConstant(RootIndex::kEmptyFixedArray));
   vobj->set(offsetof(JSPromise, reactions_or_result_), GetSmiConstant(0));
+  vobj->set(JSPromise::kContextTagOffset, GetSmiConstant(0));
   static_assert(v8::Promise::kPending == 0);
   vobj->set(offsetof(JSPromise, flags_), GetSmiConstant(0));
-  static_assert(sizeof(JSPromise) == 5 * kTaggedSize);
+  static_assert(sizeof(JSPromise) == 6 * kTaggedSize);
   for (int offset = sizeof(JSPromise);
        offset < static_cast<int>(sizeof(JSPromise)) +
                     v8::Promise::kEmbedderFieldCount * kEmbedderDataSlotSize;
diff --git a/src/objects/js-promise-inl.h b/src/objects/js-promise-inl.h
index 464b9c0712dea1c096d7993582a7399ce5ab47f7..5e07e81cfeba06c46d8400c3bfb17b535cf13b15 100644
--- a/src/objects/js-promise-inl.h
+++ b/src/objects/js-promise-inl.h
@@ -29,6 +29,12 @@ void JSPromise::set_reactions_or_result(
   reactions_or_result_.store(this, value, mode);
 }
 
+Tagged<Object> JSPromise::context_tag() const { return context_tag_.load(); }
+
+void JSPromise::set_context_tag(Tagged<Object> value, WriteBarrierMode mode) {
+  context_tag_.store(this, value, mode);
+}
+
 int JSPromise::flags() const { return flags_.load().value(); }
 
 void JSPromise::set_flags(int value) {
diff --git a/src/objects/js-promise.h b/src/objects/js-promise.h
index b76ea162d9f7b7e5a7a828d189181b0d52d31349..1071da0394a9604dd5b77866817a7802dcf76ecf 100644
--- a/src/objects/js-promise.h
+++ b/src/objects/js-promise.h
@@ -37,6 +37,10 @@ V8_OBJECT class JSPromise : public JSObjectWithEmbedderSlots {
       Tagged<UnionOf<PromiseReaction, JSAny>> value,
       WriteBarrierMode mode = UPDATE_WRITE_BARRIER);
 
+  inline Tagged<Object> context_tag() const;
+  inline void set_context_tag(Tagged<Object> value,
+                              WriteBarrierMode mode = UPDATE_WRITE_BARRIER);
+
   inline int flags() const;
   inline void set_flags(int value);
 
@@ -113,9 +117,16 @@ V8_OBJECT class JSPromise : public JSObjectWithEmbedderSlots {
   // Smi 0 terminated list of PromiseReaction objects in case the JSPromise
   // was not settled yet, otherwise the result.
   TaggedMember<UnionOf<PromiseReaction, JSAny>> reactions_or_result_;
+  // The context tag (workerd Promise Context Tagging extension).
+  TaggedMember<Object> context_tag_;
   // SmiTagged<JSPromiseFlags>.
   TaggedMember<Smi> flags_;
 
+  // Back-compat offset constant for the workerd Promise Context Tagging
+  // extension. Defined after the class body, like JSRegExp::kFlagsOffset etc.
+  static const int kContextTagOffset;
+
+
  private:
   // https://tc39.es/ecma262/#sec-triggerpromisereactions
   static Handle<Object> TriggerPromiseReactions(Isolate* isolate,
@@ -127,6 +138,9 @@ V8_OBJECT class JSPromise : public JSObjectWithEmbedderSlots {
 V8_OBJECT class JSPromiseConstructor : public JSFunctionWithPrototype {
 } V8_OBJECT_END;
 
+inline constexpr int JSPromise::kContextTagOffset =
+    offsetof(JSPromise, context_tag_);
+
 }  // namespace internal
 }  // namespace v8
 
diff --git a/src/objects/js-promise.tq b/src/objects/js-promise.tq
index 2e498caeb9a4dea6f4ef361ffe6547769e749874..227d693af230ef14dbc3f90f955701717e710fa2 100644
--- a/src/objects/js-promise.tq
+++ b/src/objects/js-promise.tq
@@ -35,6 +35,7 @@ extern class JSPromise extends JSObjectWithEmbedderSlots {
   // Smi 0 terminated list of PromiseReaction objects in case the JSPromise was
   // not settled yet, otherwise the result.
   reactions_or_result: Zero|PromiseReaction|JSAny;
+  context_tag: Object;
   flags: SmiTagged<JSPromiseFlags>;
 }
 
diff --git a/src/profiler/heap-snapshot-generator.cc b/src/profiler/heap-snapshot-generator.cc
index 887dd8c60c24a708743f35d774d0b30dca1d8e7e..2b04039b8d5731a64ba68af4f0cacc645064e093 100644
--- a/src/profiler/heap-snapshot-generator.cc
+++ b/src/profiler/heap-snapshot-generator.cc
@@ -2378,6 +2378,8 @@ void V8HeapExplorer::ExtractJSPromiseReferences(HeapEntry* entry,
   SetInternalReference(entry, "reactions_or_result",
                        promise->reactions_or_result(),
                        offsetof(JSPromise, reactions_or_result_));
+  SetInternalReference(entry, "context_tag", promise->context_tag(),
+                       JSPromise::kContextTagOffset);
 }
 
 void V8HeapExplorer::ExtractJSGeneratorObjectReferences(
diff --git a/src/runtime/runtime-promise.cc b/src/runtime/runtime-promise.cc
index 169d48bd4567686a94cc2dbca54daab3e73ad658..618dbb2f0d6e946c34f0271ec25b143cc0d7338f 100644
--- a/src/runtime/runtime-promise.cc
+++ b/src/runtime/runtime-promise.cc
@@ -236,5 +236,41 @@ RUNTIME_FUNCTION(Runtime_ConstructSuppressedError) {
   return *result;
 }
 
+RUNTIME_FUNCTION(Runtime_PromiseContextInit) {
+  HandleScope scope(isolate);
+  DCHECK_EQ(1, args.length());
+  if (!isolate->has_promise_context_tag()) {
+    args.at<JSPromise>(0)->set_context_tag(Smi::zero());
+  } else {
+    CHECK(!IsUndefined(isolate->promise_context_tag()));
+    args.at<JSPromise>(0)->set_context_tag(isolate->promise_context_tag());
+  }
+  return ReadOnlyRoots(isolate).undefined_value();
+}
+
+RUNTIME_FUNCTION(Runtime_PromiseContextCheck) {
+  HandleScope scope(isolate);
+  DCHECK_EQ(1, args.length());
+
+  Handle<JSPromise> promise = args.at<JSPromise>(0);
+
+  // If promise.context_tag() is strict equal to isolate.promise_context_tag(),
+  // or if the promise being checked does not have a context tag, we'll just
+  // return promise directly.
+  Tagged<Object> obj = promise->context_tag();
+  if (obj == Smi::zero() || obj == isolate->promise_context_tag()) {
+    return *promise;
+  }
+
+  // Otherwise we defer to the PromiseCrossContextCallback. If the callback
+  // has not been set, then it should just return the same promise back here.
+  Handle<JSPromise> result;
+  ASSIGN_RETURN_FAILURE_ON_EXCEPTION(isolate, result,
+                                     isolate->RunPromiseCrossContextCallback(
+                                         isolate->native_context(), promise));
+
+  return *result;
+}
+
 }  // namespace internal
 }  // namespace v8
diff --git a/src/runtime/runtime.h b/src/runtime/runtime.h
index b99a90b85566859592c02a6033562a251dfefc04..b2f9134e05d4c8a2a5857e033216aef467921db9 100644
--- a/src/runtime/runtime.h
+++ b/src/runtime/runtime.h
@@ -439,18 +439,20 @@ constexpr bool CanTriggerGC(T... properties) {
   F(StrictNotEqual, 2, 1)                  \
   F(ReferenceEqual, 2, 1)
 
-#define FOR_EACH_INTRINSIC_PROMISE(F, I) \
-  F(EnqueueMicrotask, 1, 1)              \
-  F(PromiseHookAfter, 1, 1)              \
-  F(PromiseHookBefore, 1, 1)             \
-  F(PromiseHookInit, 2, 1)               \
-  F(PromiseRejectEventFromStack, 2, 1)   \
-  F(PromiseRevokeReject, 1, 1)           \
-  F(RejectPromise, 3, 1)                 \
-  F(ResolvePromise, 2, 1)                \
-  F(ConstructSuppressedError, 3, 1)      \
-  F(ConstructAggregateErrorHelper, 4, 1) \
-  F(ConstructInternalAggregateErrorHelper, -1 /* <= 5*/, 1)
+#define FOR_EACH_INTRINSIC_PROMISE(F, I)                    \
+  F(EnqueueMicrotask, 1, 1)                                 \
+  F(PromiseHookAfter, 1, 1)                                 \
+  F(PromiseHookBefore, 1, 1)                                \
+  F(PromiseHookInit, 2, 1)                                  \
+  F(PromiseRejectEventFromStack, 2, 1)                      \
+  F(PromiseRevokeReject, 1, 1)                              \
+  F(RejectPromise, 3, 1)                                    \
+  F(ResolvePromise, 2, 1)                                   \
+  F(ConstructSuppressedError, 3, 1)                         \
+  F(ConstructAggregateErrorHelper, 4, 1)                    \
+  F(ConstructInternalAggregateErrorHelper, -1 /* <= 5*/, 1) \
+  F(PromiseContextInit, 1, 1)                               \
+  F(PromiseContextCheck, 1, 1)
 
 #define FOR_EACH_INTRINSIC_PROXY(F, I) \
   F(CheckProxyGetSetTrapResult, 2, 1)  \
