From c14d05513814dc04b3c8e32facf812358e882df5 Mon Sep 17 00:00:00 2001
From: Felix Hanau <felix@cloudflare.com>
Date: Sun, 9 Jul 2023 18:46:20 -0400
Subject: Enable V8 shared linkage


diff --git a/BUILD.bazel b/BUILD.bazel
index 5aeac1aa629fcb2879a33e242bf96f5a8fe1ff13..bdf7568ff0ac11d2393fc398b9f4fb84b936bd07 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -1484,7 +1484,6 @@ filegroup(
         "src/execution/futex-emulation.h",
         "src/execution/interrupts-scope.cc",
         "src/execution/interrupts-scope.h",
-        "src/execution/isolate.cc",
         "src/execution/isolate.h",
         "src/execution/isolate-data.h",
         "src/execution/isolate-inl.h",
@@ -3758,6 +3757,10 @@ filegroup(
         "src/snapshot/snapshot-empty.cc",
         "src/snapshot/static-roots-gen.cc",
         "src/snapshot/static-roots-gen.h",
+        # file moved here to make dynamic linkage possible. Keeping it in the v8 base causes issues
+        # with dynamic linking as it depends on symbols only defined in the snapshot (or through a
+        # placeholder when building the snapshot itself).
+        "src/execution/isolate.cc",
     ],
 )
 
@@ -4213,6 +4216,8 @@ v8_library(
     name = "v8",
     srcs = [
         ":v8_inspector_files",
+        # file moved here to make dynamic linkage possible.
+        "src/execution/isolate.cc",
     ] + select({
         ":is_not_v8_enable_turbofan": [
             # With Turbofan disabled, we only include the stubbed-out API.
diff --git a/bazel/defs.bzl b/bazel/defs.bzl
index f8dfd9b5eb9c65547db3b17814c84e4fbc312d07..ad8bd53dcc19c50144a6e94ed899983e014e0033 100644
--- a/bazel/defs.bzl
+++ b/bazel/defs.bzl
@@ -297,7 +297,6 @@ def v8_library(
             copts = copts + default.copts,
             linkopts = linkopts + default.linkopts,
             alwayslink = 1,
-            linkstatic = 1,
             **kwargs
         )
 
@@ -316,7 +315,6 @@ def v8_library(
             copts = copts + default.copts + ENABLE_I18N_SUPPORT_DEFINES,
             linkopts = linkopts + default.linkopts,
             alwayslink = 1,
-            linkstatic = 1,
             **kwargs
         )
 
@@ -336,7 +334,6 @@ def v8_library(
             copts = copts + default.copts,
             linkopts = linkopts + default.linkopts,
             alwayslink = 1,
-            linkstatic = 1,
             **kwargs
         )
 
