From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Felix Hanau <felix@cloudflare.com>
Date: Wed, 8 Mar 2023 16:15:31 -0500
Subject: Fix v8 code_generator imports


diff --git a/BUILD.bazel b/BUILD.bazel
index 9730021294563b213e0e601a98d4b03d1ab9f324..5639f547fab898427fe574ee50eb767e590422df 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -3459,6 +3459,11 @@ py_binary(
     deps = [
         requirement("jinja2"),
     ],
+    # The code_generator complains about not finding pdl.py, so amend the list
+    # of include directories. Not sure why this is needed only with bazel v6.
+    imports = [
+        "third_party/inspector_protocol/",
+    ],
 )
 
 genrule(
