From 159185ee2f8897f6ea666d5dd3120625678afb46 Mon Sep 17 00:00:00 2001
From: Felix Hanau <felix@cloudflare.com>
Date: Sun, 9 Jul 2023 18:46:20 -0400
Subject: Enable V8 shared linkage


diff --git a/BUILD.bazel b/BUILD.bazel
index f714fdb21097394d87f29f692fdbaa597a2f0145..4d4c716d6605dbd6e0b2395bf1e7f86b7a4df946 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -1493,7 +1493,6 @@ filegroup(
         "src/execution/futex-emulation.h",
         "src/execution/interrupts-scope.cc",
         "src/execution/interrupts-scope.h",
-        "src/execution/isolate.cc",
         "src/execution/isolate.h",
         "src/execution/isolate-data.h",
         "src/execution/isolate-inl.h",
@@ -3758,6 +3757,10 @@ filegroup(
         "src/snapshot/snapshot-empty.cc",
         "src/snapshot/static-roots-gen.cc",
         "src/snapshot/static-roots-gen.h",
+        # file moved here to make dynamic linkage possible. Keeping it in the v8 base causes issues
+        # with dynamic linking as it depends on symbols only defined in the snapshot (or through a
+        # placeholder when building the snapshot itself).
+        "src/execution/isolate.cc",
     ],
 )
 
@@ -4213,6 +4216,8 @@ v8_library(
     name = "v8",
     srcs = [
         ":v8_inspector_files",
+        # file moved here to make dynamic linkage possible.
+        "src/execution/isolate.cc",
     ] + select({
         ":is_not_v8_enable_turbofan": [
             # With Turbofan disabled, we only include the stubbed-out API.
diff --git a/bazel/defs.bzl b/bazel/defs.bzl
index 7c587522187c26326b4598149842c7e792cdcfce..a19a81d8c01fe24aab7966605bc5b6932fc277f6 100644
--- a/bazel/defs.bzl
+++ b/bazel/defs.bzl
@@ -295,7 +295,6 @@ def v8_library(
             copts = copts + default.copts,
             linkopts = linkopts + default.linkopts,
             alwayslink = 1,
-            linkstatic = 1,
             **kwargs
         )
 
@@ -314,7 +313,6 @@ def v8_library(
             copts = copts + default.copts + ENABLE_I18N_SUPPORT_DEFINES,
             linkopts = linkopts + default.linkopts,
             alwayslink = 1,
-            linkstatic = 1,
             **kwargs
         )
 
@@ -334,7 +332,6 @@ def v8_library(
             copts = copts + default.copts,
             linkopts = linkopts + default.linkopts,
             alwayslink = 1,
-            linkstatic = 1,
             **kwargs
         )
 
