From 0260f2e0b910db7342d9eed13e2aa5a80e8fbee1 Mon Sep 17 00:00:00 2001
From: Felix Hanau <felix@cloudflare.com>
Date: Sun, 9 Jul 2023 18:46:20 -0400
Subject: Enable V8 shared linkage


diff --git a/BUILD.bazel b/BUILD.bazel
index 35c8ec1656a0874c182d420954c7b1101dd6b1f7..83d0a7de3af130c1b95198bffb798a1f342a8053 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -1500,7 +1500,6 @@ filegroup(
         "src/execution/futex-emulation.h",
         "src/execution/interrupts-scope.cc",
         "src/execution/interrupts-scope.h",
-        "src/execution/isolate.cc",
         "src/execution/isolate.h",
         "src/execution/isolate-data.h",
         "src/execution/isolate-inl.h",
@@ -3758,6 +3757,10 @@ filegroup(
         "src/snapshot/snapshot-empty.cc",
         "src/snapshot/static-roots-gen.cc",
         "src/snapshot/static-roots-gen.h",
+        # file moved here to make dynamic linkage possible. Keeping it in the v8 base causes issues
+        # with dynamic linking as it depends on symbols only defined in the snapshot (or through a
+        # placeholder when building the snapshot itself).
+        "src/execution/isolate.cc",
     ],
 )
 
@@ -4214,6 +4217,8 @@ v8_library(
     name = "v8",
     srcs = [
         ":v8_inspector_files",
+        # file moved here to make dynamic linkage possible.
+        "src/execution/isolate.cc",
     ] + select({
         ":is_not_v8_enable_turbofan": [
             # With Turbofan disabled, we only include the stubbed-out API.
diff --git a/bazel/defs.bzl b/bazel/defs.bzl
index 58c5a0273769e0df997e9e92bd4274f3cb8f2fc6..32a36debd591db53f5c7dfcbdae4080e78647fd1 100644
--- a/bazel/defs.bzl
+++ b/bazel/defs.bzl
@@ -295,7 +295,6 @@ def v8_library(
             copts = copts + default.copts,
             linkopts = linkopts + default.linkopts,
             alwayslink = 1,
-            linkstatic = 1,
             **kwargs
         )
 
@@ -314,7 +313,6 @@ def v8_library(
             copts = copts + default.copts + ENABLE_I18N_SUPPORT_DEFINES,
             linkopts = linkopts + default.linkopts,
             alwayslink = 1,
-            linkstatic = 1,
             **kwargs
         )
 
@@ -334,7 +332,6 @@ def v8_library(
             copts = copts + default.copts,
             linkopts = linkopts + default.linkopts,
             alwayslink = 1,
-            linkstatic = 1,
             **kwargs
         )
 
