From b98dca40e565211ef22f7e55dd969ac2b81f2c69 Mon Sep 17 00:00:00 2001
From: Felix Hanau <felix@cloudflare.com>
Date: Sun, 9 Jul 2023 18:46:20 -0400
Subject: [PATCH 11/19] Enable V8 shared linkage

---
 BUILD.bazel    | 7 ++++++-
 bazel/defs.bzl | 3 ---
 2 files changed, 6 insertions(+), 4 deletions(-)

diff --git a/BUILD.bazel b/BUILD.bazel
index be82fc2e20a..fab5eb2419f 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -1421,7 +1421,6 @@ filegroup(
         "src/execution/futex-emulation.h",
         "src/execution/interrupts-scope.cc",
         "src/execution/interrupts-scope.h",
-        "src/execution/isolate.cc",
         "src/execution/isolate.h",
         "src/execution/isolate-data.h",
         "src/execution/isolate-inl.h",
@@ -3555,6 +3554,10 @@ filegroup(
         "src/snapshot/snapshot-empty.cc",
         "src/snapshot/static-roots-gen.cc",
         "src/snapshot/static-roots-gen.h",
+        # file moved here to make dynamic linkage possible. Keeping it in the v8 base causes issues
+        # with dynamic linking as it depends on symbols only defined in the snapshot (or through a
+        # placeholder when building the snapshot itself).
+        "src/execution/isolate.cc",
     ],
 )
 
@@ -4009,6 +4012,8 @@ v8_library(
     name = "v8",
     srcs = [
         ":v8_inspector_files",
+        # file moved here to make dynamic linkage possible.
+        "src/execution/isolate.cc",
     ] + select({
         ":is_not_v8_enable_turbofan": [
             # With Turbofan disabled, we only include the stubbed-out API.
diff --git a/bazel/defs.bzl b/bazel/defs.bzl
index 7053ca31b99..8f47a6477e8 100644
--- a/bazel/defs.bzl
+++ b/bazel/defs.bzl
@@ -293,7 +293,6 @@ def v8_library(
             copts = copts + default.copts,
             linkopts = linkopts + default.linkopts,
             alwayslink = 1,
-            linkstatic = 1,
             **kwargs
         )
 
@@ -312,7 +311,6 @@ def v8_library(
             copts = copts + default.copts + ENABLE_I18N_SUPPORT_DEFINES,
             linkopts = linkopts + default.linkopts,
             alwayslink = 1,
-            linkstatic = 1,
             **kwargs
         )
 
@@ -332,7 +330,6 @@ def v8_library(
             copts = copts + default.copts,
             linkopts = linkopts + default.linkopts,
             alwayslink = 1,
-            linkstatic = 1,
             **kwargs
         )
 
-- 
2.34.1

