From d45c01570f9d35750a9b3b179ae26761c3ecd9eb Mon Sep 17 00:00:00 2001
From: James M Snell <jasnell@gmail.com>
Date: Thu, 18 Jan 2024 10:19:14 -0800
Subject: Always enable continuation preserved data in the build


diff --git a/BUILD.bazel b/BUILD.bazel
index 83d0a7de3af130c1b95198bffb798a1f342a8053..bb12025ec61e330798f6956daea7e850b6530d1b 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -430,6 +430,7 @@ v8_config(
         "V8_ADVANCED_BIGINT_ALGORITHMS",
         "V8_CONCURRENT_MARKING",
         "V8_ENABLE_SPARKPLUG",
+        "V8_ENABLE_CONTINUATION_PRESERVED_EMBEDDER_DATA",
     ] + select({
         "@v8//bazel/config:is_debug": [
             "DEBUG",
