From 343b1d596e42f73c502b7cd1bc80c8dcdf0628aa Mon Sep 17 00:00:00 2001
From: James M Snell <jasnell@gmail.com>
Date: Thu, 18 Jan 2024 10:19:14 -0800
Subject: Always enable continuation preserved data in the build


diff --git a/BUILD.bazel b/BUILD.bazel
index 4d4c716d6605dbd6e0b2395bf1e7f86b7a4df946..be98c14965b086fca876cc38388773b8b9d1121d 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -430,6 +430,7 @@ v8_config(
         "V8_ADVANCED_BIGINT_ALGORITHMS",
         "V8_CONCURRENT_MARKING",
         "V8_ENABLE_SPARKPLUG",
+        "V8_ENABLE_CONTINUATION_PRESERVED_EMBEDDER_DATA",
     ] + select({
         "@v8//bazel/config:is_debug": [
             "DEBUG",
