From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: James M Snell <jasnell@gmail.com>
Date: Thu, 18 Jan 2024 10:19:14 -0800
Subject: Always enable continuation preserved data in the build


diff --git a/BUILD.bazel b/BUILD.bazel
index a1823f5e7a88dc0417083f3d0c14003028aaa235..d78132d07eddba8afa0084eeb0f06f6861417a0a 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -430,6 +430,7 @@ v8_config(
         "V8_ADVANCED_BIGINT_ALGORITHMS",
         "V8_CONCURRENT_MARKING",
         "V8_ENABLE_SPARKPLUG",
+        "V8_ENABLE_CONTINUATION_PRESERVED_EMBEDDER_DATA",
     ] + select({
         "@v8//bazel/config:is_debug": [
             "DEBUG",
