From 17f4495249eff3c12e172d100147cf3564adc45f Mon Sep 17 00:00:00 2001
From: James M Snell <jasnell@gmail.com>
Date: Thu, 18 Jan 2024 10:19:14 -0800
Subject: Always enable continuation preserved data in the build


diff --git a/BUILD.bazel b/BUILD.bazel
index b9c58851d0e1dc879888da3cc5edd9fd5bcb0fae..2576134d57c089ebaebd0d0567388d3ae26ea336 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -430,6 +430,7 @@ v8_config(
         "V8_ADVANCED_BIGINT_ALGORITHMS",
         "V8_CONCURRENT_MARKING",
         "V8_ENABLE_SPARKPLUG",
+        "V8_ENABLE_CONTINUATION_PRESERVED_EMBEDDER_DATA",
     ] + select({
         "@v8//bazel/config:is_debug": [
             "DEBUG",
