From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Erik Corry <erikcorry@chromium.org>
Date: Mon, 2 Dec 2024 14:16:37 +0100
Subject: Add another slot in the isolate for embedder

Signed-off-by: James M Snell <jsnell@cloudflare.com>

diff --git a/include/v8-internal.h b/include/v8-internal.h
index 22a0ab55d18ebb483f9e7ce92ac4a25aac628536..f44067012b304b21fac8b98195dd5cdd402d125c 100644
--- a/include/v8-internal.h
+++ b/include/v8-internal.h
@@ -1013,7 +1013,7 @@ class Internals {
   // AccessorInfo::data and InterceptorInfo::data field.
   static const int kCallbackInfoDataOffset = 1 * kApiTaggedSize;
 
-  static const uint32_t kNumIsolateDataSlots = 4;
+  static const uint32_t kNumIsolateDataSlots = 6;
   static const int kStackGuardSize = 8 * kApiSystemPointerSize;
   static const int kNumberOfBooleanFlags = 6;
   static const int kErrorMessageParamSize = 1;
