From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Erik Corry <erikcorry@chromium.org>
Date: Mon, 2 Dec 2024 14:16:37 +0100
Subject: Add another slot in the isolate for embedder

Signed-off-by: James M Snell <jsnell@cloudflare.com>

diff --git a/include/v8-internal.h b/include/v8-internal.h
index 9a39f8ede92bad9acf1c390cde2048d7a9eb6a88..10dcd7b24007cf5bfb75cc0e33c777f4964fb76c 100644
--- a/include/v8-internal.h
+++ b/include/v8-internal.h
@@ -918,7 +918,7 @@ class Internals {
   static const int kExternalTwoByteRepresentationTag = 0x02;
   static const int kExternalOneByteRepresentationTag = 0x0a;
 
-  static const uint32_t kNumIsolateDataSlots = 4;
+  static const uint32_t kNumIsolateDataSlots = 5;
   static const int kStackGuardSize = 8 * kApiSystemPointerSize;
   static const int kNumberOfBooleanFlags = 6;
   static const int kErrorMessageParamSize = 1;
