From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Erik Corry <erikcorry@chromium.org>
Date: Mon, 2 Dec 2024 14:16:37 +0100
Subject: Add another slot in the isolate for embedder

Signed-off-by: James M Snell <jsnell@cloudflare.com>

diff --git a/include/v8-internal.h b/include/v8-internal.h
index 281f669323caee35e72e1dfa0b64c63c3ff15dcb..cf12b527d62a717479e4c83dd59e4659660aa22a 100644
--- a/include/v8-internal.h
+++ b/include/v8-internal.h
@@ -1018,7 +1018,7 @@ class Internals {
   // AccessorInfo::data and InterceptorInfo::data field.
   static const int kCallbackInfoDataOffset = 1 * kApiTaggedSize;
 
-  static const uint32_t kNumIsolateDataSlots = 4;
+  static const uint32_t kNumIsolateDataSlots = 6;
   static const int kStackGuardSize = 8 * kApiSystemPointerSize;
   static const int kNumberOfBooleanFlags = 6;
   static const int kErrorMessageParamSize = 1;
