From 53626a47eed7ca39c323c708ee6d088680249c5b Mon Sep 17 00:00:00 2001
From: James M Snell <jasnell@gmail.com>
Date: Tue, 9 Jul 2024 08:37:05 -0700
Subject: Update illegal invocation error message in v8


diff --git a/src/common/message-template.h b/src/common/message-template.h
index 2928f78dcbd551dfd14fb7c01670a9b282c29b41..503ce8ac38a26be6aa4abb048e19c5ddce76e2b4 100644
--- a/src/common/message-template.h
+++ b/src/common/message-template.h
@@ -112,7 +112,9 @@ namespace internal {
     "First argument to % must not be a regular expression")                    \
   T(FunctionBind, "Bind must be called on a function")                         \
   T(GeneratorRunning, "Generator is already running")                          \
-  T(IllegalInvocation, "Illegal invocation")                                   \
+  T(IllegalInvocation,                                                         \
+    "Illegal invocation: function called with incorrect `this` reference. "    \
+    "See https://developers.cloudflare.com/workers/observability/errors/#illegal-invocation-errors for details.") \
   T(ImmutablePrototypeSet,                                                     \
     "Immutable prototype object '%' cannot have their prototype set")          \
   T(ImportAttributesDuplicateKey, "Import attribute has duplicate key '%'")    \
