From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Erik Corry <erikcorry@chromium.org>
Date: Mon, 2 Dec 2024 14:16:37 +0100
Subject: Add another slot in the isolate for embedder


diff --git a/include/v8-internal.h b/include/v8-internal.h
index a6cf2f89352d7b584f568502b0b93eb973ed93b4..fc7666e6ffb0644feb108313e6bd73b49337090f 100644
--- a/include/v8-internal.h
+++ b/include/v8-internal.h
@@ -866,7 +866,7 @@ class Internals {
   static const int kExternalTwoByteRepresentationTag = 0x02;
   static const int kExternalOneByteRepresentationTag = 0x0a;
 
-  static const uint32_t kNumIsolateDataSlots = 4;
+  static const uint32_t kNumIsolateDataSlots = 5;
   static const int kStackGuardSize = 8 * kApiSystemPointerSize;
   static const int kNumberOfBooleanFlags = 6;
   static const int kErrorMessageParamSize = 1;
