From c4887f798d55850a858fbb046a526194cf267c11 Mon Sep 17 00:00:00 2001
From: James M Snell <jasnell@gmail.com>
Date: Tue, 9 Jul 2024 08:37:05 -0700
Subject: Update illegal invocation error message in v8


diff --git a/src/common/message-template.h b/src/common/message-template.h
index 20e4ce1b57d947bc80d10eb1baf05f495b244505..25eec5c469261f655a226782b7bad55892108ff4 100644
--- a/src/common/message-template.h
+++ b/src/common/message-template.h
@@ -112,7 +112,9 @@ namespace internal {
     "First argument to % must not be a regular expression")                    \
   T(FunctionBind, "Bind must be called on a function")                         \
   T(GeneratorRunning, "Generator is already running")                          \
-  T(IllegalInvocation, "Illegal invocation")                                   \
+  T(IllegalInvocation,                                                         \
+    "Illegal invocation: function called with incorrect `this` reference. "    \
+    "See https://developers.cloudflare.com/workers/observability/errors/#illegal-invocation-errors for details.") \
   T(ImmutablePrototypeSet,                                                     \
     "Immutable prototype object '%' cannot have their prototype set")          \
   T(ImportAssertDeprecated,                                                    \
