From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Erik Corry <erikcorry@chromium.org>
Date: Mon, 2 Dec 2024 14:16:37 +0100
Subject: Add another slot in the isolate for embedder


diff --git a/include/v8-internal.h b/include/v8-internal.h
index b56e67375dab7bffd1e083f49bf1faf0583af4bc..22818cd5c6f36d6591fab4ac2d0184a9c291c531 100644
--- a/include/v8-internal.h
+++ b/include/v8-internal.h
@@ -834,7 +834,7 @@ class Internals {
   static const int kExternalTwoByteRepresentationTag = 0x02;
   static const int kExternalOneByteRepresentationTag = 0x0a;
 
-  static const uint32_t kNumIsolateDataSlots = 4;
+  static const uint32_t kNumIsolateDataSlots = 5;
   static const int kStackGuardSize = 8 * kApiSystemPointerSize;
   static const int kNumberOfBooleanFlags = 6;
   static const int kErrorMessageParamSize = 1;
