From 5065e34454ff23a4434e373e11d4d46579f65f0e Mon Sep 17 00:00:00 2001
From: James M Snell <jasnell@gmail.com>
Date: Mon, 16 Sep 2024 09:56:04 -0700
Subject: Implement cross-request context promise resolve handling


diff --git a/include/v8-callbacks.h b/include/v8-callbacks.h
index 6add89ac1837df03a71539801ab513e098e73722..ad153256a06f16fc89cee9144716ba6bfaaec08c 100644
--- a/include/v8-callbacks.h
+++ b/include/v8-callbacks.h
@@ -463,6 +463,27 @@ using PromiseCrossContextCallback = MaybeLocal<Promise> (*)(Local<Context> conte
                                                             Local<Promise> promise,
                                                             Local<Object> tag);
 
+/**
+ * PromiseCrossContextResolveCallback is called when resolving or rejecting a
+ * pending promise whose context tag is not strictly equal to the isolate's
+ * current promise context tag. The callback is called with the promise to be
+ * resolved, it's context tag, and a function that when called, causes the
+ * reactions to the resolved promise to be enqueued. The idea is that the
+ * embedder sets this callback in the case it needs to defer the actual
+ * scheduling of the reactions to the given promise to a later time.
+ * Importantly, when this callback is invoked, the state of the promise
+ * should have already been updated. We're simply possibly deferring the
+ * enqueue of the reactions to the promise.
+ */
+using PromiseCrossContextResolveCallback = Maybe<void> (*)(
+    v8::Isolate* isolate,
+    Local<Value> tag,
+    Local<Data> reactions,
+    Local<Value> argument,
+    std::function<void(v8::Isolate* isolate,
+                       Local<Data> reactions,
+                       Local<Value> argument)> callback);
+
 }  // namespace v8
 
 #endif  // INCLUDE_V8_ISOLATE_CALLBACKS_H_
diff --git a/include/v8-isolate.h b/include/v8-isolate.h
index 96b7d9feb20d4ef1b5ba2da56d9635688da2457c..825ead73595d445e0cf27df4c4c43de5a31632e3 100644
--- a/include/v8-isolate.h
+++ b/include/v8-isolate.h
@@ -1733,6 +1733,8 @@ class V8_EXPORT Isolate {
 
   class PromiseContextScope;
   void SetPromiseCrossContextCallback(PromiseCrossContextCallback callback);
+  void SetPromiseCrossContextResolveCallback(
+      PromiseCrossContextResolveCallback callback);
 
   Isolate() = delete;
   ~Isolate() = delete;
diff --git a/src/api/api.cc b/src/api/api.cc
index 06895d14a0ab93b8a9f16b761744d8d8986bdf97..cbe6b3153a4e165154680ae1670d5d1fd83633ba 100644
--- a/src/api/api.cc
+++ b/src/api/api.cc
@@ -12247,12 +12247,17 @@ void Isolate::SetPromiseCrossContextCallback(PromiseCrossContextCallback callbac
   isolate->set_promise_cross_context_callback(callback);
 }
 
+void Isolate::SetPromiseCrossContextResolveCallback(PromiseCrossContextResolveCallback callback) {
+  i::Isolate* isolate = reinterpret_cast<i::Isolate*>(this);
+  isolate->set_promise_cross_context_resolve_callback(callback);
+}
+
 Isolate::PromiseContextScope::PromiseContextScope(Isolate* isolate, v8::Local<v8::Object> tag)
     : isolate_(reinterpret_cast<i::Isolate*>(isolate)) {
   DCHECK(!isolate_->has_promise_context_tag());
   DCHECK(!tag.IsEmpty());
   i::Handle<i::Object> handle = Utils::OpenHandle(*tag);
-  isolate_->set_promise_context_tag(*handle);
+  isolate_->set_promise_context_tag(handle);
 }
 
 Isolate::PromiseContextScope::~PromiseContextScope() {
diff --git a/src/builtins/promise-abstract-operations.tq b/src/builtins/promise-abstract-operations.tq
index 6142e0450dac646ccebd3fc7c5168f9b42b4d0b5..9acc20825711c87aa3fe07916cddf208e0963b9f 100644
--- a/src/builtins/promise-abstract-operations.tq
+++ b/src/builtins/promise-abstract-operations.tq
@@ -23,6 +23,9 @@ extern transitioning runtime PromiseRejectEventFromStack(
 
 extern transitioning runtime PromiseContextCheck(
     implicit context: Context)(JSPromise): JSPromise;
+
+extern transitioning runtime PromiseResolveContextCheck(
+    implicit context: Context)(JSPromise): JSAny;
 }
 
 // https://tc39.es/ecma262/#sec-promise-abstract-operations
@@ -239,7 +242,8 @@ transitioning builtin RejectPromise(
   // the runtime handle this operation, which greatly reduces
   // the complexity here and also avoids a couple of back and
   // forth between JavaScript and C++ land.
-  if (IsIsolatePromiseHookEnabledOrDebugIsActiveOrHasAsyncEventDelegate(
+  if (ToBoolean(runtime::PromiseResolveContextCheck(promise)) ||
+      IsIsolatePromiseHookEnabledOrDebugIsActiveOrHasAsyncEventDelegate(
           promiseHookFlags) ||
       !promise.HasHandler()) {
     // 7. If promise.[[PromiseIsHandled]] is false, perform
diff --git a/src/builtins/promise-resolve.tq b/src/builtins/promise-resolve.tq
index 202180adbbae91a689a667c40d20b4b1b9cb6edd..c93ac5905d7b349d1c59e9fa86b48662313ea1c3 100644
--- a/src/builtins/promise-resolve.tq
+++ b/src/builtins/promise-resolve.tq
@@ -96,7 +96,9 @@ transitioning builtin ResolvePromise(
   // We also let the runtime handle it if promise == resolution.
   // We can use pointer comparison here, since the {promise} is guaranteed
   // to be a JSPromise inside this function and thus is reference comparable.
-  if (IsIsolatePromiseHookEnabledOrDebugIsActiveOrHasAsyncEventDelegate() ||
+
+  if (ToBoolean(runtime::PromiseResolveContextCheck(promise)) ||
+      IsIsolatePromiseHookEnabledOrDebugIsActiveOrHasAsyncEventDelegate() ||
       TaggedEqual(promise, resolution))
     deferred {
       return runtime::ResolvePromise(promise, resolution);
diff --git a/src/execution/isolate-inl.h b/src/execution/isolate-inl.h
index 9e0969460fd33d8cc893e04d3418e1cbed01d6a0..aa2b1cea6b7b6c0c91108776e79b04cc8be64b9d 100644
--- a/src/execution/isolate-inl.h
+++ b/src/execution/isolate-inl.h
@@ -134,26 +134,35 @@ bool Isolate::is_execution_terminating() {
          i::ReadOnlyRoots(this).termination_exception();
 }
 
-Tagged<Object> Isolate::promise_context_tag() {
-  return promise_context_tag_;
+Handle<Object> Isolate::promise_context_tag() {
+  return root_handle(RootIndex::kPromiseContextTag);
 }
 
 bool Isolate::has_promise_context_tag() {
-  return promise_context_tag_ != ReadOnlyRoots(this).the_hole_value();
+  return heap()->promise_context_tag() != ReadOnlyRoots(this).the_hole_value();
 }
 
 void Isolate::clear_promise_context_tag() {
-  set_promise_context_tag(ReadOnlyRoots(this).the_hole_value());
+  heap()->set_promise_context_tag(ReadOnlyRoots(this).the_hole_value());
 }
 
-void Isolate::set_promise_context_tag(Tagged<Object> tag) {
-  promise_context_tag_ = tag;
+void Isolate::set_promise_context_tag(Handle<Object> tag) {
+  heap()->set_promise_context_tag(*tag);
 }
 
 void Isolate::set_promise_cross_context_callback(PromiseCrossContextCallback callback) {
   promise_cross_context_callback_ = callback;
 }
 
+void Isolate::set_promise_cross_context_resolve_callback(
+    PromiseCrossContextResolveCallback callback) {
+  promise_cross_context_resolve_callback_ = callback;
+}
+
+bool Isolate::has_promise_context_resolve_callback() {
+  return promise_cross_context_resolve_callback_ != nullptr;
+}
+
 #ifdef DEBUG
 Tagged<Object> Isolate::VerifyBuiltinsResult(Tagged<Object> result) {
   if (is_execution_terminating() && !v8_flags.strict_termination_checks) {
diff --git a/src/execution/isolate.cc b/src/execution/isolate.cc
index d093aad3a30c8a6c04e1c7cf2b2ea0d8afc03e5b..4f094be2cc98be530b219fecfe6c3e9da81ef29d 100644
--- a/src/execution/isolate.cc
+++ b/src/execution/isolate.cc
@@ -599,8 +599,6 @@ void Isolate::Iterate(RootVisitor* v, ThreadLocalTop* thread) {
                       FullObjectSlot(&thread->pending_message_));
   v->VisitRootPointer(Root::kStackRoots, nullptr,
                       FullObjectSlot(&thread->context_));
-  v->VisitRootPointer(Root::kStackRoots, nullptr,
-                      FullObjectSlot(&promise_context_tag_));
 
   for (v8::TryCatch* block = thread->try_catch_handler_; block != nullptr;
        block = block->next_) {
@@ -7516,5 +7514,27 @@ MaybeHandle<JSPromise> Isolate::RunPromiseCrossContextCallback(Handle<NativeCont
   return v8::Utils::OpenHandle(*result);
 }
 
+Maybe<void> Isolate::RunPromiseCrossContextResolveCallback(v8::Isolate* isolate,
+                                                           Handle<JSObject> tag,
+                                                           Handle<Object> reactions,
+                                                           Handle<Object> argument,
+                                                           PromiseReaction::Type type) {
+  CHECK(promise_cross_context_resolve_callback_ != nullptr);
+  return promise_cross_context_resolve_callback_(
+      isolate,
+      v8::Utils::ToLocal(tag),
+      v8::Utils::ToLocal(reactions),
+      v8::Utils::ToLocal(argument),
+      [type](v8::Isolate* isolate,
+             v8::Local<v8::Data> reactions,
+             v8::Local<v8::Value> argument) {
+    JSPromise::ContinueTriggerPromiseReactions(
+        reinterpret_cast<Isolate*>(isolate),
+        Utils::OpenHandle(*reactions),
+        Utils::OpenHandle(*argument),
+        type);
+  });
+}
+
 }  // namespace internal
 }  // namespace v8
diff --git a/src/execution/isolate.h b/src/execution/isolate.h
index 7de738063bdddead3762aded57fbbd388eef950f..f6736d735b243966c2b219f12457ed29508fe3ab 100644
--- a/src/execution/isolate.h
+++ b/src/execution/isolate.h
@@ -42,6 +42,7 @@
 #include "src/objects/contexts.h"
 #include "src/objects/debug-objects.h"
 #include "src/objects/js-objects.h"
+#include "src/objects/promise.h"
 #include "src/objects/tagged.h"
 #include "src/runtime/runtime.h"
 #include "src/sandbox/code-pointer-table.h"
@@ -2266,13 +2267,22 @@ class V8_EXPORT_PRIVATE Isolate final : private HiddenFactory {
                                        v8::ExceptionContext callback_kind);
   void SetExceptionPropagationCallback(ExceptionPropagationCallback callback);
 
-  inline Tagged<Object> promise_context_tag();
+  inline Handle<Object> promise_context_tag();
   inline bool has_promise_context_tag();
   inline void clear_promise_context_tag();
-  inline void set_promise_context_tag(Tagged<Object> tag);
+  inline void set_promise_context_tag(Handle<Object> tag);
   inline void set_promise_cross_context_callback(PromiseCrossContextCallback callback);
+  inline void set_promise_cross_context_resolve_callback(
+      PromiseCrossContextResolveCallback callback);
   MaybeHandle<JSPromise> RunPromiseCrossContextCallback(Handle<NativeContext> context,
                                                         Handle<JSPromise> promise);
+  Maybe<void> RunPromiseCrossContextResolveCallback(v8::Isolate* isolate,
+                                                    Handle<JSObject> tag,
+                                                    Handle<Object> reactions,
+                                                    Handle<Object> argument,
+                                                    PromiseReaction::Type type);
+
+  inline bool has_promise_context_resolve_callback();
 
 #ifdef V8_ENABLE_WASM_SIMD256_REVEC
   void set_wasm_revec_verifier_for_test(
@@ -2786,9 +2796,10 @@ class V8_EXPORT_PRIVATE Isolate final : private HiddenFactory {
   int current_thread_counter_ = 0;
 #endif
 
-  Tagged<Object> promise_context_tag_;
-  PromiseCrossContextCallback promise_cross_context_callback_;
+  PromiseCrossContextCallback promise_cross_context_callback_ = nullptr;
+  PromiseCrossContextResolveCallback promise_cross_context_resolve_callback_ = nullptr;
   bool in_promise_cross_context_callback_ = false;
+  bool in_promise_cross_context_resolve_callback_ = false;
 
   class PromiseCrossContextCallbackScope;
 
diff --git a/src/heap/factory.cc b/src/heap/factory.cc
index 1d631dde7b3234f6d3639a8f49410e908126f6d6..baafc83265c8386437b8399e1e50050c1109965f 100644
--- a/src/heap/factory.cc
+++ b/src/heap/factory.cc
@@ -4427,18 +4427,17 @@ Handle<JSPromise> Factory::NewJSPromiseWithoutHook() {
   Handle<JSPromise> promise =
       Cast<JSPromise>(NewJSObject(isolate()->promise_function()));
   DisallowGarbageCollection no_gc;
-  Tagged<JSPromise> raw = *promise;
-  raw->set_reactions_or_result(Smi::zero(), SKIP_WRITE_BARRIER);
+  promise->set_reactions_or_result(Smi::zero(), SKIP_WRITE_BARRIER);
   if (!isolate()->has_promise_context_tag()) {
-    raw->set_context_tag(Smi::zero(), SKIP_WRITE_BARRIER);
+    promise->set_context_tag(Smi::zero(), SKIP_WRITE_BARRIER);
   } else {
-    raw->set_context_tag(isolate()->promise_context_tag());
+    promise->set_context_tag(*isolate()->promise_context_tag());
   }
 
-  raw->set_flags(0);
+  promise->set_flags(0);
   // TODO(v8) remove once embedder data slots are always zero-initialized.
   InitEmbedderFields(*promise, Smi::zero());
-  DCHECK_EQ(raw->GetEmbedderFieldCount(), v8::Promise::kEmbedderFieldCount);
+  DCHECK_EQ(promise->GetEmbedderFieldCount(), v8::Promise::kEmbedderFieldCount);
   return promise;
 }
 
diff --git a/src/objects/js-promise.h b/src/objects/js-promise.h
index afb40bd86464535ce0b7ed649cd0b09ea78b3d26..e1e953affdc617747164e655d1063f906a03ac43 100644
--- a/src/objects/js-promise.h
+++ b/src/objects/js-promise.h
@@ -81,6 +81,11 @@ class JSPromise
   static_assert(v8::Promise::kFulfilled == 1);
   static_assert(v8::Promise::kRejected == 2);
 
+  static void ContinueTriggerPromiseReactions(Isolate* isolate,
+                                              DirectHandle<Object> reactions,
+                                              DirectHandle<Object> argument,
+                                              PromiseReaction::Type type);
+
  private:
   // ES section #sec-triggerpromisereactions
   static Handle<Object> TriggerPromiseReactions(Isolate* isolate,
diff --git a/src/objects/objects.cc b/src/objects/objects.cc
index eda5636e187c368ec849c6ab204fe49c335a691e..c67cc54200ebb717d3b4c962218858e220cc59aa 100644
--- a/src/objects/objects.cc
+++ b/src/objects/objects.cc
@@ -4866,6 +4866,23 @@ Handle<Object> JSPromise::Fulfill(DirectHandle<JSPromise> promise,
   // 6. Set promise.[[PromiseState]] to "fulfilled".
   promise->set_status(Promise::kFulfilled);
 
+  Handle<Object> obj(promise->context_tag(), isolate);
+  bool needs_promise_context_switch =
+      !(*obj == Smi::zero() ||
+        obj.is_identical_to(isolate->promise_context_tag()) ||
+        !isolate->has_promise_context_resolve_callback());
+  if (needs_promise_context_switch) {
+    if (isolate->RunPromiseCrossContextResolveCallback(
+        reinterpret_cast<v8::Isolate*>(isolate),
+        Cast<JSObject>(obj),
+        reactions,
+        value,
+        PromiseReaction::kFulfill).IsNothing()) {
+      return {};
+    }
+    return isolate->factory()->undefined_value();
+  }
+
   // 7. Return TriggerPromiseReactions(reactions, value).
   return TriggerPromiseReactions(isolate, reactions, value,
                                  PromiseReaction::kFulfill);
@@ -4921,6 +4938,23 @@ Handle<Object> JSPromise::Reject(Handle<JSPromise> promise,
     isolate->ReportPromiseReject(promise, reason, kPromiseRejectWithNoHandler);
   }
 
+  Handle<Object> obj(promise->context_tag(), isolate);
+  bool needs_promise_context_switch =
+      !(*obj == Smi::zero() ||
+        obj.is_identical_to(isolate->promise_context_tag()) ||
+        !isolate->has_promise_context_resolve_callback());
+  if (needs_promise_context_switch) {
+    if (isolate->RunPromiseCrossContextResolveCallback(
+        reinterpret_cast<v8::Isolate*>(isolate),
+        Cast<JSObject>(obj),
+        reactions,
+        reason,
+        PromiseReaction::kReject).IsNothing()) {
+      return {};
+    }
+    return isolate->factory()->undefined_value();
+  }
+
   // 8. Return TriggerPromiseReactions(reactions, reason).
   return TriggerPromiseReactions(isolate, reactions, reason,
                                  PromiseReaction::kReject);
@@ -5023,6 +5057,15 @@ MaybeHandle<Object> JSPromise::Resolve(Handle<JSPromise> promise,
 }
 
 // static
+
+void JSPromise::ContinueTriggerPromiseReactions(
+    Isolate* isolate,
+    DirectHandle<Object> reactions,
+    DirectHandle<Object> argument,
+    PromiseReaction::Type type) {
+  TriggerPromiseReactions(isolate, reactions, argument, type);
+}
+
 Handle<Object> JSPromise::TriggerPromiseReactions(
     Isolate* isolate, DirectHandle<Object> reactions,
     DirectHandle<Object> argument, PromiseReaction::Type type) {
diff --git a/src/roots/roots.h b/src/roots/roots.h
index 47e0ef53ee79901873b8dd48e7ca19465cea1868..2657550796b0a022b5efb9830ac0896ba5ca9d39 100644
--- a/src/roots/roots.h
+++ b/src/roots/roots.h
@@ -406,7 +406,8 @@ class RootVisitor;
   V(FunctionTemplateInfo, error_stack_getter_fun_template,                  \
     ErrorStackGetterSharedFun)                                              \
   V(FunctionTemplateInfo, error_stack_setter_fun_template,                  \
-    ErrorStackSetterSharedFun)
+    ErrorStackSetterSharedFun)                                              \
+  V(Object, promise_context_tag, PromiseContextTag)
 
 // Entries in this list are limited to Smis and are not visited during GC.
 #define SMI_ROOT_LIST(V)                                                       \
diff --git a/src/runtime/runtime-promise.cc b/src/runtime/runtime-promise.cc
index 4fbdf90319c5fc3b3e6f60aaf1b12a33c5f537ea..f58e6ce6173695f8caf77ba1aa6cf8de8ccf3ca2 100644
--- a/src/runtime/runtime-promise.cc
+++ b/src/runtime/runtime-promise.cc
@@ -133,8 +133,10 @@ RUNTIME_FUNCTION(Runtime_RejectPromise) {
   Handle<JSPromise> promise = args.at<JSPromise>(0);
   Handle<Object> reason = args.at(1);
   DirectHandle<Boolean> debug_event = args.at<Boolean>(2);
-  return *JSPromise::Reject(promise, reason,
+  Handle<Object> result = JSPromise::Reject(promise, reason,
                             Object::BooleanValue(*debug_event, isolate));
+  RETURN_FAILURE_IF_EXCEPTION(isolate);
+  return *result;
 }
 
 RUNTIME_FUNCTION(Runtime_ResolvePromise) {
@@ -222,8 +224,8 @@ RUNTIME_FUNCTION(Runtime_PromiseContextInit) {
   if (!isolate->has_promise_context_tag()) {
     args.at<JSPromise>(0)->set_context_tag(Smi::zero());
   } else {
-    CHECK(!IsUndefined(isolate->promise_context_tag()));
-    args.at<JSPromise>(0)->set_context_tag(isolate->promise_context_tag());
+    CHECK(!IsUndefined(*isolate->promise_context_tag()));
+    args.at<JSPromise>(0)->set_context_tag(*isolate->promise_context_tag());
   }
   return ReadOnlyRoots(isolate).undefined_value();
 }
@@ -237,8 +239,8 @@ RUNTIME_FUNCTION(Runtime_PromiseContextCheck) {
   // If promise.context_tag() is strict equal to isolate.promise_context_tag(),
   // or if the promise being checked does not have a context tag, we'll just return
   // promise directly.
-  Tagged<Object> obj = promise->context_tag();
-  if (obj == Smi::zero() || obj == isolate->promise_context_tag()) {
+  Handle<Object> obj(promise->context_tag(), isolate);
+  if (*obj == Smi::zero() || obj.is_identical_to(isolate->promise_context_tag())) {
     return *promise;
   }
 
@@ -251,5 +253,23 @@ RUNTIME_FUNCTION(Runtime_PromiseContextCheck) {
   return *result;
 }
 
+RUNTIME_FUNCTION(Runtime_PromiseResolveContextCheck) {
+  HandleScope scope(isolate);
+  DCHECK_EQ(1, args.length());
+  Handle<JSPromise> promise = args.at<JSPromise>(0);
+  // If promise.context_tag() is strict equal to isolate.promise_context_tag(),
+  // or if the promise being checked does not have a context tag, or if the
+  // resolve callback has not been set, we'll just return false here to indicate
+  // that the default handling should be used.
+  Handle<Object> obj(promise->context_tag(), isolate);
+  if (*obj == Smi::zero() ||
+      obj.is_identical_to(isolate->promise_context_tag()) ||
+      !isolate->has_promise_context_resolve_callback()) {
+    return isolate->heap()->ToBoolean(false);
+  }
+
+  return isolate->heap()->ToBoolean(true);
+}
+
 }  // namespace internal
 }  // namespace v8
diff --git a/src/runtime/runtime.h b/src/runtime/runtime.h
index d23446cbeee65b50cab694dc60743b4ab7f9c1d7..a9e5f7e459442f3c9cb4b3d1e9491da9c87ab114 100644
--- a/src/runtime/runtime.h
+++ b/src/runtime/runtime.h
@@ -405,7 +405,8 @@ namespace internal {
   F(ConstructAggregateErrorHelper, 4, 1) \
   F(ConstructInternalAggregateErrorHelper, -1 /* <= 5*/, 1) \
   F(PromiseContextInit, 1, 1)            \
-  F(PromiseContextCheck, 1, 1)
+  F(PromiseContextCheck, 1, 1)           \
+  F(PromiseResolveContextCheck, 1, 1)
 
 #define FOR_EACH_INTRINSIC_PROXY(F, I) \
   F(CheckProxyGetSetTrapResult, 2, 1)  \
