From bd3333d53de62140f33b4b8ba8656f2162da2c47 Mon Sep 17 00:00:00 2001
From: James M Snell <jasnell@gmail.com>
Date: Tue, 9 Jul 2024 08:37:05 -0700
Subject: Update illegal invocation error message in v8


diff --git a/src/common/message-template.h b/src/common/message-template.h
index 5c7ffea18f6dd9976735213fff2e117225350f26..22e3f17402390dcdc723e2a9d4a56e048567f20b 100644
--- a/src/common/message-template.h
+++ b/src/common/message-template.h
@@ -112,7 +112,9 @@ namespace internal {
     "First argument to % must not be a regular expression")                    \
   T(FunctionBind, "Bind must be called on a function")                         \
   T(GeneratorRunning, "Generator is already running")                          \
-  T(IllegalInvocation, "Illegal invocation")                                   \
+  T(IllegalInvocation,                                                         \
+    "Illegal invocation: function called with incorrect `this` reference. "    \
+    "See https://developers.cloudflare.com/workers/observability/errors/#illegal-invocation-errors for details.") \
   T(ImmutablePrototypeSet,                                                     \
     "Immutable prototype object '%' cannot have their prototype set")          \
   T(ImportAssertDeprecated,                                                    \
