From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Erik Corry <erikcorry@chromium.org>
Date: Mon, 7 Apr 2025 12:32:30 +0200
Subject: Remove unneded -latomic linker flag

Signed-off-by: James M Snell <jsnell@cloudflare.com>

diff --git a/bazel/defs.bzl b/bazel/defs.bzl
index 853fad73496a70f59475810cdd4d809862487ccf..adaf444c04b5055bdc4b8c8f42cde308bb2f180b 100644
--- a/bazel/defs.bzl
+++ b/bazel/defs.bzl
@@ -211,7 +211,7 @@ def _default_args():
                 "Shell32.lib",
             ],
             "@v8//bazel/config:is_macos": ["-pthread"],
-            "//conditions:default": ["-Wl,--no-as-needed -ldl -latomic -pthread"],
+            "//conditions:default": ["-Wl,--no-as-needed -ldl -pthread"],
         }) + select({
             ":should_add_rdynamic": ["-rdynamic"],
             "//conditions:default": [],
