From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Erik Corry <erikcorry@chromium.org>
Date: Mon, 7 Apr 2025 12:32:30 +0200
Subject: Remove unneded -latomic linker flag

Signed-off-by: James M Snell <jsnell@cloudflare.com>

diff --git a/bazel/defs.bzl b/bazel/defs.bzl
index c041e4176b918dd33c7c25579f36dc8c63e2c8c0..b22b7f7520461db8520cfd754ed6082d94544b4a 100644
--- a/bazel/defs.bzl
+++ b/bazel/defs.bzl
@@ -212,7 +212,7 @@ def _default_args():
                 "Shell32.lib",
             ],
             "@v8//bazel/config:is_macos": ["-pthread"],
-            "//conditions:default": ["-Wl,--no-as-needed -ldl -latomic -pthread"],
+            "//conditions:default": ["-Wl,--no-as-needed -ldl -pthread"],
         }) + select({
             ":should_add_rdynamic": ["-rdynamic"],
             "//conditions:default": [],
