From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Felix Hanau <felix@cloudflare.com>
Date: Sun, 9 Jul 2023 18:46:20 -0400
Subject: Enable V8 shared linkage


diff --git a/BUILD.bazel b/BUILD.bazel
index cbef1193f598eb1bbb95eee9857ad70efb2f0d3f..d42347b15fa323835c7b2e65346548ae21573ae2 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -1366,6 +1366,7 @@ filegroup(
         "src/builtins/constants-table-builder.cc",
         "src/builtins/constants-table-builder.h",
         "src/builtins/data-view-ops.h",
+        "src/builtins/profile-data-reader.cc",
         "src/builtins/profile-data-reader.h",
         "src/codegen/aligned-slot-allocator.cc",
         "src/codegen/aligned-slot-allocator.h",
@@ -1547,7 +1548,6 @@ filegroup(
         "src/execution/futex-emulation.h",
         "src/execution/interrupts-scope.cc",
         "src/execution/interrupts-scope.h",
-        "src/execution/isolate.cc",
         "src/execution/isolate.h",
         "src/execution/isolate-data.h",
         "src/execution/isolate-inl.h",
@@ -3068,7 +3068,6 @@ filegroup(
 filegroup(
     name = "v8_compiler_files",
     srcs = [
-        "src/builtins/profile-data-reader.h",
         "src/compiler/access-builder.cc",
         "src/compiler/access-builder.h",
         "src/compiler/access-info.cc",
@@ -3642,8 +3641,6 @@ filegroup(
         "src/builtins/growable-fixed-array-gen.cc",
         "src/builtins/growable-fixed-array-gen.h",
         "src/builtins/number-builtins-reducer-inl.h",
-        "src/builtins/profile-data-reader.cc",
-        "src/builtins/profile-data-reader.h",
         "src/builtins/setup-builtins-internal.cc",
         "src/builtins/torque-csa-header-includes.h",
         "src/codegen/turboshaft-builtins-assembler-inl.h",
@@ -3923,6 +3920,7 @@ filegroup(
         "src/snapshot/snapshot-empty.cc",
         "src/snapshot/static-roots-gen.cc",
         "src/snapshot/static-roots-gen.h",
+        "src/execution/isolate.cc",
     ],
 )
 
@@ -4033,6 +4031,10 @@ filegroup(
     name = "noicu/snapshot_files",
     srcs = [
         "src/init/setup-isolate-deserialize.cc",
+        # Not a snapshot file per se, but depends on symbols only present in the snapshot (or
+        # through a placeholder when building the snapshot itself); having it here makes using
+        # shared linkage for the main v8 target possible.
+        "src/execution/isolate.cc",
     ] + select({
         "@v8//bazel/config:v8_target_arm": [
             "google3/snapshots/arm/noicu/embedded.S",
@@ -4050,6 +4052,7 @@ filegroup(
     name = "icu/snapshot_files",
     srcs = [
         "src/init/setup-isolate-deserialize.cc",
+        "src/execution/isolate.cc",
     ] + select({
         "@v8//bazel/config:v8_target_arm": [
             "google3/snapshots/arm/icu/embedded.S",
diff --git a/bazel/defs.bzl b/bazel/defs.bzl
index 55384199d235d28fba18eac4a6c0c51e1e7675a4..be179a1fee5a81a022be63b93a8a0e3c8b4f68ca 100644
--- a/bazel/defs.bzl
+++ b/bazel/defs.bzl
@@ -297,7 +297,6 @@ def v8_library(
             copts = copts + default.copts,
             linkopts = linkopts + default.linkopts,
             alwayslink = 1,
-            linkstatic = 1,
             **kwargs
         )
 
@@ -316,7 +315,6 @@ def v8_library(
             copts = copts + default.copts + ENABLE_I18N_SUPPORT_DEFINES,
             linkopts = linkopts + default.linkopts,
             alwayslink = 1,
-            linkstatic = 1,
             **kwargs
         )
 
@@ -336,7 +334,6 @@ def v8_library(
             copts = copts + default.copts,
             linkopts = linkopts + default.linkopts,
             alwayslink = 1,
-            linkstatic = 1,
             **kwargs
         )
 
