From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Felix Hanau <felix@cloudflare.com>
Date: Sun, 8 Jun 2025 16:40:37 -0400
Subject: Port V8_USE_ZLIB support

Change-Id: Icfedf3e90522f1ff5037517a39a5f0e3d44abace

diff --git a/BUILD.bazel b/BUILD.bazel
index 71db68eef5cce90ec721dabdef446aee224bce9a..42ed7c61f0a5bcb5316e624e8c5b29c19d355970 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -162,6 +162,11 @@ v8_flag(name = "v8_enable_verify_predictable")
 
 v8_flag(name = "v8_enable_test_features")
 
+v8_flag(
+    name = "v8_use_zlib",
+    default = True,
+)
+
 v8_flag(name = "v8_wasm_random_fuzzers")
 
 v8_flag(
@@ -533,6 +538,7 @@ v8_config(
         "v8_enable_vtunejit": "ENABLE_VTUNE_JIT_INTERFACE",
         "v8_enable_undefined_double": "V8_ENABLE_UNDEFINED_DOUBLE",
         "v8_use_host_cpu_arm_features": "V8_USE_HOST_CPU_ARM_FEATURES",
+        "v8_use_zlib": "V8_USE_ZLIB",
     },
     defines = [
         "GOOGLE3",
@@ -4676,6 +4682,8 @@ v8_library(
         "@highway//:hwy",
         "@fast_float",
         "@simdutf",
+        "@zlib",
+        "@zlib//:compression_utils_portable",
     ],
 )
 
diff --git a/src/deoptimizer/frame-translation-builder.cc b/src/deoptimizer/frame-translation-builder.cc
index 497f11e5cd2c20b33b86a71615fb12e48c32048f..b2bbe6aca39bec6299551be0932eda89034775e8 100644
--- a/src/deoptimizer/frame-translation-builder.cc
+++ b/src/deoptimizer/frame-translation-builder.cc
@@ -11,7 +11,7 @@
 #include "src/objects/fixed-array-inl.h"
 
 #ifdef V8_USE_ZLIB
-#include "third_party/zlib/google/compression_utils_portable.h"
+#include <google/compression_utils_portable.h>
 #endif  // V8_USE_ZLIB
 
 namespace v8 {
diff --git a/src/objects/deoptimization-data.cc b/src/objects/deoptimization-data.cc
index db3b6ee16e17c8eb0c338e76054dc59c08967c24..be26237d335be4cde756f147940227262f6917ed 100644
--- a/src/objects/deoptimization-data.cc
+++ b/src/objects/deoptimization-data.cc
@@ -14,7 +14,7 @@
 #include "src/objects/shared-function-info.h"
 
 #ifdef V8_USE_ZLIB
-#include "third_party/zlib/google/compression_utils_portable.h"
+#include <google/compression_utils_portable.h>
 #endif  // V8_USE_ZLIB
 
 namespace v8 {
diff --git a/src/snapshot/snapshot-utils.cc b/src/snapshot/snapshot-utils.cc
index ea04fcec0ecebd409d1ed9839dd0db3ac73160a9..e67eb478a0f49fa2a6c4f338e86a7e6a5e6d686f 100644
--- a/src/snapshot/snapshot-utils.cc
+++ b/src/snapshot/snapshot-utils.cc
@@ -7,7 +7,7 @@
 #include "src/base/sanitizer/msan.h"
 
 #ifdef V8_USE_ZLIB
-#include "third_party/zlib/zlib.h"
+#include <zlib.h>
 #endif
 
 namespace v8 {
