From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Felix Hanau <felix@cloudflare.com>
Date: Sun, 8 Jun 2025 16:39:03 -0400
Subject: Port concurrent mksnapshot support

Change-Id: I57c8158ff5d624e5379e6b072f27ac7a40419522

diff --git a/BUILD.bazel b/BUILD.bazel
index d49cfb18585538cb3a557ef484a16ae04305c0ca..afdc0516e43449c3d99115f174520ac77e0b7200 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -120,6 +120,11 @@ v8_flag(name = "v8_enable_hugepage")
 
 v8_flag(name = "v8_enable_fast_mksnapshot")
 
+v8_flag(
+    name = "v8_enable_concurrent_mksnapshot",
+    default = True,
+)
+
 v8_flag(name = "v8_enable_future")
 
 # NOTE: Transitions are not recommended in library targets:
@@ -4405,6 +4410,13 @@ v8_mksnapshot(
             "--no-turbo-verify-allocation",
         ],
         "//conditions:default": [],
+    }) + select({
+        ":is_v8_enable_concurrent_mksnapshot": [
+            "--concurrent-builtin-generation",
+            # Use all the cores for concurrent builtin generation.
+            "--concurrent-turbofan-max-threads=0",
+        ],
+        "//conditions:default": [],
     }) + select({
         ":is_v8_enable_snapshot_code_comments": ["--code-comments"],
         "//conditions:default": [],
