From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yagiz Nizipli <yagiz@nizipli.com>
Date: Thu, 19 Jun 2025 10:05:19 -0400
Subject: Modify where to look for dragonbox


diff --git a/BUILD.bazel b/BUILD.bazel
index eb0124cc974928f6ac1baba7ea6e12173b4d069a..1fbf951d325322da19071c96585f2d74c6a2c2d8 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -4204,13 +4204,8 @@ filegroup(
 
 v8_library(
     name = "lib_dragonbox",
-    srcs = ["third_party/dragonbox/src/include/dragonbox/dragonbox.h"],
-    hdrs = [
-        "third_party/dragonbox/src/include/dragonbox/dragonbox.h",
-    ],
-    includes = [
-        "third_party/dragonbox/src/include",
-    ],
+    srcs = [],
+    deps = ["@dragonbox"],
 )
 
 cc_library(
