From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yagiz Nizipli <yagiz@nizipli.com>
Date: Thu, 19 Jun 2025 10:05:19 -0400
Subject: Modify where to look for dragonbox


diff --git a/BUILD.bazel b/BUILD.bazel
index 23665c89fd04db517ccf0ab8dccb5346cd5cac45..63682dc1766c6036fd8f10ff717589a85bb25c23 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -4151,14 +4151,9 @@ filegroup(
 )
 
 v8_library(
-  name = "lib_dragonbox",
-  srcs = ["third_party/dragonbox/src/include/dragonbox/dragonbox.h"],
-  hdrs = [
-    "third_party/dragonbox/src/include/dragonbox/dragonbox.h",
-  ],
-  includes = [
-    "third_party/dragonbox/src/include",
-  ],
+    name = "lib_dragonbox",
+    srcs = [],
+    deps = ["@dragonbox"]
 )
 
 cc_library(
