From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yagiz Nizipli <yagiz@nizipli.com>
Date: Thu, 19 Jun 2025 10:05:19 -0400
Subject: Modify where to look for dragonbox


diff --git a/BUILD.bazel b/BUILD.bazel
index 30e16e4ec102c04a4a4193214b10702e1fd5eb70..91ead9e0f732879aee7cbb5b155217a5b7ab9afb 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -4189,13 +4189,8 @@ filegroup(
 
 v8_library(
     name = "lib_dragonbox",
-    srcs = ["third_party/dragonbox/src/include/dragonbox/dragonbox.h"],
-    hdrs = [
-        "third_party/dragonbox/src/include/dragonbox/dragonbox.h",
-    ],
-    includes = [
-        "third_party/dragonbox/src/include",
-    ],
+    srcs = [],
+    deps = ["@dragonbox"],
 )
 
 cc_library(
