From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yagiz Nizipli <yagiz@nizipli.com>
Date: Thu, 19 Jun 2025 10:05:19 -0400
Subject: Modify where to look for dragonbox


diff --git a/BUILD.bazel b/BUILD.bazel
index 369e3f975beacf9aea41f05aae919554a8944a61..b3ae97e40f7fd77ba9122ad35421759540a0a100 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -4089,14 +4089,9 @@ filegroup(
 )
 
 v8_library(
-  name = "lib_dragonbox",
-  srcs = ["third_party/dragonbox/src/include/dragonbox/dragonbox.h"],
-  hdrs = [
-    "third_party/dragonbox/src/include/dragonbox/dragonbox.h",
-  ],
-  includes = [
-    "third_party/dragonbox/src/include",
-  ],
+    name = "lib_dragonbox",
+    srcs = [],
+    deps = ["@dragonbox"]
 )
 
 cc_library(
