From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yagiz Nizipli <yagiz@nizipli.com>
Date: Thu, 19 Jun 2025 10:05:19 -0400
Subject: Modify where to look for dragonbox


diff --git a/BUILD.bazel b/BUILD.bazel
index a714df90df89e4932f2620ab6afeb60f0d6e5e57..e255deb3845b13aec48441beef07b2ec954ea5c9 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -4199,13 +4199,8 @@ filegroup(
 
 v8_library(
     name = "lib_dragonbox",
-    srcs = ["third_party/dragonbox/src/include/dragonbox/dragonbox.h"],
-    hdrs = [
-        "third_party/dragonbox/src/include/dragonbox/dragonbox.h",
-    ],
-    includes = [
-        "third_party/dragonbox/src/include",
-    ],
+    srcs = [],
+    deps = ["@dragonbox"],
 )
 
 cc_library(
