From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yagiz Nizipli <yagiz@nizipli.com>
Date: Thu, 19 Jun 2025 10:05:19 -0400
Subject: Modify where to look for dragonbox


diff --git a/BUILD.bazel b/BUILD.bazel
index 29b68d89b4d54917c19a0de783fdea5ff3b5783f..30502729d7a58c28924fdad2d44e4b97237fda1b 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -4186,13 +4186,8 @@ filegroup(
 
 v8_library(
     name = "lib_dragonbox",
-    srcs = ["third_party/dragonbox/src/include/dragonbox/dragonbox.h"],
-    hdrs = [
-        "third_party/dragonbox/src/include/dragonbox/dragonbox.h",
-    ],
-    includes = [
-        "third_party/dragonbox/src/include",
-    ],
+    srcs = [],
+    deps = ["@dragonbox"],
 )
 
 cc_library(
