From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yagiz Nizipli <yagiz@nizipli.com>
Date: Thu, 19 Jun 2025 10:05:19 -0400
Subject: Modify where to look for dragonbox


diff --git a/BUILD.bazel b/BUILD.bazel
index e6d32a244ed32d278b2c88cadbc40406c1a4d928..933969d80185900284195293408ee479fff1cf40 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -3981,14 +3981,9 @@ filegroup(
 )
 
 v8_library(
-  name = "lib_dragonbox",
-  srcs = ["third_party/dragonbox/src/include/dragonbox/dragonbox.h"],
-  hdrs = [
-    "third_party/dragonbox/src/include/dragonbox/dragonbox.h",
-  ],
-  includes = [
-    "third_party/dragonbox/src/include",
-  ],
+    name = "lib_dragonbox",
+    srcs = [],
+    deps = ["@dragonbox"]
 )
 
 cc_library(
