From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yagiz Nizipli <yagiz@nizipli.com>
Date: Thu, 19 Jun 2025 10:05:19 -0400
Subject: Modify where to look for dragonbox


diff --git a/BUILD.bazel b/BUILD.bazel
index d2f6f0234dbebc32da7f1c588aaa138e57a14128..f5b9f82ad28b2dd97969183817ba43ab13866087 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -3926,14 +3926,9 @@ filegroup(
 )
 
 v8_library(
-  name = "lib_dragonbox",
-  srcs = ["third_party/dragonbox/src/include/dragonbox/dragonbox.h"],
-  hdrs = [
-    "third_party/dragonbox/src/include/dragonbox/dragonbox.h",
-  ],
-  includes = [
-    "third_party/dragonbox/src/include",
-  ],
+    name = "lib_dragonbox",
+    srcs = [],
+    deps = ["@dragonbox"]
 )
 
 cc_library(
diff --git a/src/numbers/conversions.cc b/src/numbers/conversions.cc
index 1eec37da8d7570ebf6246c46f346774e3d8945e0..8f1b64ba81336fa806aa32cb5d66b4411e4380e6 100644
--- a/src/numbers/conversions.cc
+++ b/src/numbers/conversions.cc
@@ -23,7 +23,7 @@
 #include "src/objects/string-inl.h"
 #include "src/strings/char-predicates-inl.h"
 #include "src/utils/allocation.h"
-#include "third_party/dragonbox/src/include/dragonbox/dragonbox.h"
+#include "third_party/dragonbox/include/dragonbox/dragonbox.h"
 #include "third_party/fast_float/src/include/fast_float/fast_float.h"
 #include "third_party/fast_float/src/include/fast_float/float_common.h"
 
