From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Alex Eagle <alex@aspect.dev>
Date: Thu, 2 Oct 2025 02:17:00 +0100
Subject: bind icu to googlesource


diff --git a/BUILD.bazel b/BUILD.bazel
index 1fbf951d325322da19071c96585f2d74c6a2c2d8..a5e2035ec799f5df7367b9b8d6e49831397eef70 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -4914,7 +4914,7 @@ v8_library(
     copts = ["-Wno-implicit-fallthrough"],
     icu_deps = [
         ":icu/generated_torque_definitions_headers",
-        "//external:icu",
+        "@com_googlesource_chromium_icu//:icu",
     ] + select({
         ":is_v8_use_metagen_instance_types": [
             ":icu/metagen_instance_types_headers",
@@ -5053,7 +5053,7 @@ v8_binary(
     ],
     deps = [
         ":v8_libbase",
-        "//external:icu",
+        "@com_googlesource_chromium_icu//:icu",
     ],
 )
 
