From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Alex Eagle <alex@aspect.dev>
Date: Thu, 2 Oct 2025 02:17:00 +0100
Subject: bind icu to googlesource


diff --git a/BUILD.bazel b/BUILD.bazel
index b3ae97e40f7fd77ba9122ad35421759540a0a100..159142a018dd65f608c986cde611b0aed2df4bed 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -4618,7 +4618,7 @@ v8_library(
     copts = ["-Wno-implicit-fallthrough"],
     icu_deps = [
         ":icu/generated_torque_definitions_headers",
-        "//external:icu",
+        "@com_googlesource_chromium_icu//:icu",
     ],
     icu_srcs = [
         ":generated_regexp_special_case",
@@ -4741,7 +4741,7 @@ v8_binary(
     ],
     deps = [
         ":v8_libbase",
-        "//external:icu",
+        "@com_googlesource_chromium_icu//:icu",
     ],
 )
 
