From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Alex Eagle <alex@aspect.dev>
Date: Thu, 2 Oct 2025 02:17:00 +0100
Subject: bind icu to googlesource


diff --git a/BUILD.bazel b/BUILD.bazel
index e255deb3845b13aec48441beef07b2ec954ea5c9..56c9e18986d8004c12001a3dd445704079d05a56 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -4909,7 +4909,7 @@ v8_library(
     copts = ["-Wno-implicit-fallthrough"],
     icu_deps = [
         ":icu/generated_torque_definitions_headers",
-        "//external:icu",
+        "@com_googlesource_chromium_icu//:icu",
     ] + select({
         ":is_v8_use_metagen_instance_types": [
             ":icu/metagen_instance_types_headers",
@@ -5048,7 +5048,7 @@ v8_binary(
     ],
     deps = [
         ":v8_libbase",
-        "//external:icu",
+        "@com_googlesource_chromium_icu//:icu",
     ],
 )
 
