From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Alex Eagle <alex@aspect.dev>
Date: Thu, 2 Oct 2025 02:17:00 +0100
Subject: bind icu to googlesource


diff --git a/BUILD.bazel b/BUILD.bazel
index 607d8bb1ee30a02623798c206097eb34f748cd98..2bb1144a0247e4279010524e9e4241900880e7b0 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -4537,7 +4537,7 @@ v8_library(
     copts = ["-Wno-implicit-fallthrough"],
     icu_deps = [
         ":icu/generated_torque_definitions_headers",
-        "//external:icu",
+        "@com_googlesource_chromium_icu//:icu",
     ],
     icu_srcs = [
         ":generated_regexp_special_case",
@@ -4660,7 +4660,7 @@ v8_binary(
     ],
     deps = [
         ":v8_libbase",
-        "//external:icu",
+        "@com_googlesource_chromium_icu//:icu",
     ],
 )
 
