From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: James M Snell <jsnell@cloudflare.com>
Date: Wed, 5 Nov 2025 12:22:49 -0800
Subject: optimize ascii fast path in WriteUtf8V2

Refs: https://chromium-review.googlesource.com/c/v8/v8/+/7124103
Signed-off-by: James M Snell <jsnell@cloudflare.com>

diff --git a/src/strings/unicode-inl.h b/src/strings/unicode-inl.h
index f45cee3d7c0203ce710699668e3b280bb4f0c750..9420b3a9f08e62c7013ecff99d16da07851c2546 100644
--- a/src/strings/unicode-inl.h
+++ b/src/strings/unicode-inl.h
@@ -206,6 +206,16 @@ bool Utf8::IsValidCharacter(uchar c) {
           c != kBadChar);
 }
 
+template <>
+bool Utf8::IsAsciiOneByteString<uint8_t>(const uint8_t* buffer, size_t size) {
+  return simdutf::validate_ascii(reinterpret_cast<const char*>(buffer), size);
+}
+
+template <>
+bool Utf8::IsAsciiOneByteString<uint16_t>(const uint16_t* buffer, size_t size) {
+  return false;
+}
+
 template <typename Char>
 Utf8::EncodingResult Utf8::Encode(v8::base::Vector<const Char> string,
                                   char* buffer, size_t capacity,
@@ -221,8 +231,17 @@ Utf8::EncodingResult Utf8::Encode(v8::base::Vector<const Char> string,
   const Char* characters = string.begin();
   size_t content_capacity = capacity - write_null;
   CHECK_LE(content_capacity, capacity);
-  uint16_t last = Utf16::kNoPreviousCharacter;
   size_t read_index = 0;
+  if (kSourceIsOneByte) {
+    size_t writeable = std::min(string.size(), content_capacity);
+    // Just memcpy when possible.
+    if (writeable > 0 && Utf8::IsAsciiOneByteString(characters, writeable)) {
+      memcpy(buffer, characters, writeable);
+      read_index = writeable;
+      write_index = writeable;
+    }
+  }
+  uint16_t last = Utf16::kNoPreviousCharacter;
   for (; read_index < string.size(); read_index++) {
     Char character = characters[read_index];
 
diff --git a/src/strings/unicode.h b/src/strings/unicode.h
index ef1e717b1ea8579df7c2597c3ca19ef84110e3c4..32a0b84a8399b2a38cb6f5e721dcb9fdaaf320d0 100644
--- a/src/strings/unicode.h
+++ b/src/strings/unicode.h
@@ -212,6 +212,16 @@ class V8_EXPORT_PRIVATE Utf8 {
   // - valid code point range.
   static bool ValidateEncoding(const uint8_t* str, size_t length);
 
+  template <typename Char>
+  static bool IsAsciiOneByteString(const Char* buffer, size_t size);
+
+  template <>
+  inline bool IsAsciiOneByteString<uint8_t>(const uint8_t* buffer, size_t size);
+
+  template <>
+  inline bool IsAsciiOneByteString<uint16_t>(const uint16_t* buffer,
+                                             size_t size);
+
   // Encode the given characters as Utf8 into the provided output buffer.
   struct EncodingResult {
     size_t bytes_written;
