From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ketan Gupta <ketan@cloudflare.com>
Date: Wed, 4 Feb 2026 15:05:21 +0000
Subject: Add verify_write_barriers flag in V8's bazel config


diff --git a/BUILD.bazel b/BUILD.bazel
index 30923227154ac6a928a8938b417c12589f7b005d..ca7a16d82f8e214d80b226fb92fa9e848a732777 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -567,6 +567,7 @@ v8_config(
         "@v8//bazel/config:is_debug": [
             "DEBUG",
             "V8_ENABLE_CHECKS",
+            "V8_VERIFY_WRITE_BARRIERS",
         ],
         "//conditions:default": [],
     }) + select({
