From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ketan Gupta <ketan@cloudflare.com>
Date: Wed, 4 Feb 2026 15:05:21 +0000
Subject: Add verify_write_barriers flag in V8's bazel config


diff --git a/BUILD.bazel b/BUILD.bazel
index 8751b0e6d8cff2e7bdf8f825b4f0caa8152995d1..d8af05a91dc117de3f82adb41f4180e4d38c1f51 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -559,6 +559,7 @@ v8_config(
         "@v8//bazel/config:is_debug": [
             "DEBUG",
             "V8_ENABLE_CHECKS",
+            "V8_VERIFY_WRITE_BARRIERS",
         ],
         "//conditions:default": [],
     }) + select({
