From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ketan Gupta <ketan@cloudflare.com>
Date: Wed, 4 Feb 2026 15:05:21 +0000
Subject: Add verify_write_barriers flag in V8's bazel config


diff --git a/BUILD.bazel b/BUILD.bazel
index 159142a018dd65f608c986cde611b0aed2df4bed..d954621014d8c3b8c5d005382edee9fd936709d0 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -540,6 +540,7 @@ v8_config(
         "@v8//bazel/config:is_debug": [
             "DEBUG",
             "V8_ENABLE_CHECKS",
+            "V8_VERIFY_WRITE_BARRIERS",
         ],
         "//conditions:default": [],
     }) + select({
