From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ketan Gupta <ketan@cloudflare.com>
Date: Wed, 4 Feb 2026 15:05:21 +0000
Subject: Add verify_write_barriers flag in V8's bazel config


diff --git a/BUILD.bazel b/BUILD.bazel
index a5e2035ec799f5df7367b9b8d6e49831397eef70..47d285a961edcecd3495845375530f159b2af291 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -575,6 +575,7 @@ v8_config(
         "@v8//bazel/config:is_debug": [
             "DEBUG",
             "V8_ENABLE_CHECKS",
+            "V8_VERIFY_WRITE_BARRIERS",
         ],
         "//conditions:default": [],
     }) + select({
