From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ketan Gupta <ketan@cloudflare.com>
Date: Wed, 4 Feb 2026 15:05:21 +0000
Subject: Add verify_write_barriers flag in V8's bazel config


diff --git a/BUILD.bazel b/BUILD.bazel
index 56c9e18986d8004c12001a3dd445704079d05a56..858dd0fece9ea1979752528d5badffc5ccad5607 100644
--- a/BUILD.bazel
+++ b/BUILD.bazel
@@ -575,6 +575,7 @@ v8_config(
         "@v8//bazel/config:is_debug": [
             "DEBUG",
             "V8_ENABLE_CHECKS",
+            "V8_VERIFY_WRITE_BARRIERS",
         ],
         "//conditions:default": [],
     }) + select({
