From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ketan Gupta <ketan@cloudflare.com>
Date: Wed, 4 Feb 2026 15:48:36 +0000
Subject: Change lamba signature to get around windows build failure


diff --git a/src/objects/backing-store.cc b/src/objects/backing-store.cc
index fbf90e6a641405e298b39649be4bc2279d15719b..80a8de60b37ffa9925754442f63c09a3de92065e 100644
--- a/src/objects/backing-store.cc
+++ b/src/objects/backing-store.cc
@@ -316,7 +316,7 @@ std::unique_ptr<BackingStore> BackingStore::TryAllocateAndPartiallyCommitMemory(
   // For accounting purposes, whether a GC was necessary.
   bool did_retry = false;
 
-  auto gc_retry = [&](const std::function<bool()>& fn) {
+  auto gc_retry = [&](auto&& fn) {
     if (fn()) return true;
     // Collect garbage and retry.
     did_retry = true;
