From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ketan Gupta <ketan@cloudflare.com>
Date: Wed, 4 Feb 2026 15:48:36 +0000
Subject: Change lamba signature to get around windows build failure


diff --git a/src/objects/backing-store.cc b/src/objects/backing-store.cc
index c98ab72ad532350404f1fac0bedf3c4e83a1ab03..e43b8ea2e693b9981b18a87191ebee5b903d4872 100644
--- a/src/objects/backing-store.cc
+++ b/src/objects/backing-store.cc
@@ -322,7 +322,7 @@ std::unique_ptr<BackingStore> BackingStore::TryAllocateAndPartiallyCommitMemory(
   // For accounting purposes, whether a GC was necessary.
   bool did_retry = false;
 
-  auto gc_retry = [&](const std::function<bool()>& fn) {
+  auto gc_retry = [&](auto&& fn) {
     if (fn()) return true;
     // Collect garbage and retry.
     did_retry = true;
