From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ketan Gupta <ketan@cloudflare.com>
Date: Wed, 4 Feb 2026 15:48:36 +0000
Subject: Change lamba signature to get around windows build failure


diff --git a/src/objects/backing-store.cc b/src/objects/backing-store.cc
index 251ddd3c131866da41eb7c0ca47c790480d4d239..2bfba594a264093a72dc34c853c83863fa79cb8f 100644
--- a/src/objects/backing-store.cc
+++ b/src/objects/backing-store.cc
@@ -310,7 +310,7 @@ std::unique_ptr<BackingStore> BackingStore::TryAllocateAndPartiallyCommitMemory(
   // For accounting purposes, whether a GC was necessary.
   bool did_retry = false;
 
-  auto gc_retry = [&](const std::function<bool()>& fn) {
+  auto gc_retry = [&](auto&& fn) {
     if (fn()) return true;
     // Collect garbage and retry.
     did_retry = true;
