From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ketan Gupta <ketan@cloudflare.com>
Date: Wed, 4 Feb 2026 15:48:36 +0000
Subject: Change lamba signature to get around windows build failure


diff --git a/src/objects/backing-store.cc b/src/objects/backing-store.cc
index 5297cd116b571bb4c679c4f44a3c58781dcc712d..f9e815e3a47fe8caa1bc874d06e57d7c8b3f999f 100644
--- a/src/objects/backing-store.cc
+++ b/src/objects/backing-store.cc
@@ -321,7 +321,7 @@ std::unique_ptr<BackingStore> BackingStore::TryAllocateAndPartiallyCommitMemory(
   // For accounting purposes, whether a GC was necessary.
   bool did_retry = false;
 
-  auto gc_retry = [&](const std::function<bool()>& fn) {
+  auto gc_retry = [&](auto&& fn) {
     if (fn()) return true;
     // Collect garbage and retry.
     did_retry = true;
