From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dan Carney <dcarney@chromium.org>
Date: Tue, 17 Feb 2026 08:45:16 +0000
Subject: Return false on Object.hasOwnProperty with interceptors


diff --git a/src/objects/js-objects.cc b/src/objects/js-objects.cc
index af6cdd90b7382960d4e230fd9357467d2d155249..066ef6fb2f0524d48b162044f3b829f4f6c466f0 100644
--- a/src/objects/js-objects.cc
+++ b/src/objects/js-objects.cc
@@ -158,6 +158,9 @@ Maybe<bool> JSReceiver::HasOwnProperty(Isolate* isolate,
   if (IsJSObject(*object)) {  // Shortcut.
     PropertyKey key(isolate, name);
     LookupIterator it(isolate, object, key, LookupIterator::OWN);
+    if (it.state() == LookupIterator::INTERCEPTOR) {
+      return Just(false);
+    }
     return HasProperty(&it);
   }
 
