From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dan Carney <dcarney@chromium.org>
Date: Tue, 17 Feb 2026 08:45:16 +0000
Subject: Return false on Object.hasOwnProperty with interceptors


diff --git a/src/objects/js-objects.cc b/src/objects/js-objects.cc
index b6950f5d2eb652cd651ef181f9d9e2e135df06ed..1bca8eab0b538f885e05b13911a1ac3e5897360a 100644
--- a/src/objects/js-objects.cc
+++ b/src/objects/js-objects.cc
@@ -165,6 +165,9 @@ Maybe<bool> JSReceiver::HasOwnProperty(Isolate* isolate,
   if (IsJSObject(*object)) {  // Shortcut.
     PropertyKey key(isolate, name);
     LookupIterator it(isolate, object, key, LookupIterator::OWN);
+    if (it.state() == LookupIterator::INTERCEPTOR) {
+      return Just(false);
+    }
     return HasProperty(&it);
   }
 
