From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dan Carney <dcarney@chromium.org>
Date: Tue, 17 Feb 2026 08:45:16 +0000
Subject: Return false on Object.hasOwnProperty with interceptors


diff --git a/src/objects/js-objects.cc b/src/objects/js-objects.cc
index 15fce882065d99d1d01e5c138e651118727e9a28..76ceae5d9878df28cd4aa6780fc1ae449a9ea89b 100644
--- a/src/objects/js-objects.cc
+++ b/src/objects/js-objects.cc
@@ -158,6 +158,9 @@ Maybe<bool> JSReceiver::HasOwnProperty(Isolate* isolate,
   if (IsJSObject(*object)) {  // Shortcut.
     PropertyKey key(isolate, name);
     LookupIterator it(isolate, object, key, LookupIterator::OWN);
+    if (it.state() == LookupIterator::INTERCEPTOR) {
+      return Just(false);
+    }
     return HasProperty(&it);
   }
 
