From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Vicky Kontoura <vasiliki@cloudflare.com>
Date: Wed, 8 Apr 2026 23:52:30 +0300
Subject: Delete copy constructor of PropertyDeduplicator to fix Windows
 compilation issue


diff --git a/src/ast/ast.cc b/src/ast/ast.cc
index 46c40e0979abba02851d984dc618f7a4f9ec2997..162672109fa719645d75d67e84a009da94b97fbd 100644
--- a/src/ast/ast.cc
+++ b/src/ast/ast.cc
@@ -496,6 +496,7 @@ namespace {
 //      duplicate
 class PropertyDeduplicator {
  public:
+  PropertyDeduplicator(const PropertyDeduplicator&) = delete;
   PropertyDeduplicator(const ZonePtrList<ObjectLiteral::Property>* properties,
                        uint32_t boilerplate_properties)
       : properties_(properties),
