From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mike Aizatsky <maizatskyi@cloudflare.com>
Date: Wed, 5 Mar 2025 09:38:53 -0800
Subject: massage build.rs for workerd


diff --git a/build.rs b/build.rs
index 87b8063a4725f89d86f0f8fd06cad03ffca947eb..48f81d313686846cc817d58747d47bb264dc7565 100644
--- a/build.rs
+++ b/build.rs
@@ -75,7 +75,7 @@ fn main() {
     return;
   }
 
-  print_link_flags();
+  // print_link_flags();
 
   // Don't attempt rebuild but link
   if is_trybuild {
@@ -115,9 +115,7 @@ fn main() {
     return;
   }
 
-  print_prebuilt_src_binding_path();
-
-  download_static_lib_binaries();
+  build_binding();
 }
 
 fn acquire_lock() -> LockFile {
@@ -150,7 +148,7 @@ fn build_binding() {
   let bindings = bindgen::Builder::default()
     .header("src/binding.hpp")
     .parse_callbacks(Box::new(bindgen::CargoCallbacks::new()))
-    .clang_args(["-x", "c++", "-std=c++20", "-Iv8/include", "-I."])
+    .clang_args(["-x", "c++", "-std=c++20", "-DV8_TYPED_ARRAY_MAX_SIZE_IN_HEAP=64", "-I../v8/include", "-I."])
     .clang_args(args)
     .generate_cstr(true)
     .rustified_enum(".*UseCounterFeature")
@@ -163,7 +161,7 @@ fn build_binding() {
     .generate()
     .expect("Unable to generate bindings");
 
-  let out_path = build_dir().join("gn_out").join("src_binding.rs");
+  let out_path = out_dir().join("src_binding.rs");
   println!(
     "cargo:rustc-env=RUSTY_V8_SRC_BINDING_PATH={}",
     out_path.display()
@@ -448,18 +446,19 @@ fn static_lib_dir() -> PathBuf {
   build_dir().join("gn_out").join("obj")
 }
 
-fn build_dir() -> PathBuf {
+fn out_dir() -> PathBuf {
   let cwd = env::current_dir().unwrap();
-
-  // target/debug//build/rusty_v8-d9e5a424d4f96994/out/
   let out_dir = env::var_os("OUT_DIR").expect(
     "The 'OUT_DIR' environment is not set (it should be something like \
      'target/debug/rusty_v8-{hash}').",
-  );
+    );
   let out_dir_abs = cwd.join(out_dir);
+  out_dir_abs
+}
 
+fn build_dir() -> PathBuf {
   // This would be target/debug or target/release
-  out_dir_abs
+  out_dir()
     .parent()
     .unwrap()
     .parent()
